<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="In depth tutorial about HicAggR functions.
">
<title>HicAggR - In depth tutorial • HicAggR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="HicAggR - In depth tutorial">
<meta property="og:description" content="In depth tutorial about HicAggR functions.
">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">HicAggR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.7</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/HicAggR.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/InDepth.html">HicAggR - In depth tutorial</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/CuvierLab/HicAggR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="InDepth_files/kePrint-0.0.1/kePrint.js"></script><link href="InDepth_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>HicAggR - In depth tutorial</h1>
                        <h4 data-toc-skip class="author">Nicolas
Chanard, David Depierre, Robel A. Tesfaye, Stéphane Schaak &amp; Olivier
Cuvier</h4>
            
            <h4 data-toc-skip class="date">2024-04-29</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/CuvierLab/HicAggR/blob/HEAD/vignettes/InDepth.Rmd" class="external-link"><code>vignettes/InDepth.Rmd</code></a></small>
      <div class="d-none name"><code>InDepth.Rmd</code></div>
    </div>

    
    
<style>
body {
text-align: justify}
</style>
<style type="text/css">
pre {
    max-height: 300px;
    overflow-y: auto;
}
pre[class] {
    max-height: 300px;
}
</style>
<style type="text/css">
.scroll-100 {
    max-height: 100px;
    overflow-y: auto;
    background-color: inherit;
}
</style>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>HicAggR package provides a set of tools for the analysis of genomic
3D conformation data (HiC). It is designed around Aggregation Peak
Analysis (APA), with the intent to provide easy-to-use sets of functions
that would allow users to integrate 1D-genomics data with 3D-genomics
data. This package does not perform TAD calling, loop calling, or
compartment analysis. There are other packages available for these
analyses (eg: TopDom (from CRAN), HiTC (bioconductor), HiCDOC
(bioconductor)…). The package offers however tools to perform downstream
analyses on called loops,or make use of called TADs to explore
interactions within loops or explore intra/inter-TAD interactions. To
this end, pixels of interest surrounding called loops or genomic
features are extracted from whole HiC matrices (as submatrices), abiding
to TAD and/or compartmental constraints in order to give user an insight
into the interaction patterns of features of interest or called
loops.</p>
<div class="section level3">
<h3 id="typical-workflow">Typical workflow:<a class="anchor" aria-label="anchor" href="#typical-workflow"></a>
</h3>
<p>The straight-forward workflow with this package would be:</p>
<ol style="list-style-type: decimal">
<li>to import genomic coordinates of features of interest (such as,
enhancers, CTCF binding sites, genes) as <code>GRanges</code>
</li>
<li>apply distance constraints or 3D structural constraints (like TADs)
to form all potential couples between features of interest under the
applied contraints. (Enhancer-promoter, CTCF-CTCF etc.)</li>
<li>import 3D-genomics data (.hic, bedpe, h5, cool/mcool formats).</li>
<li>perform matrix balancing and distance normalization if
necessary.</li>
<li>extract submatrices of the pre-formed potential couples from the
3D-genomics Data.</li>
<li>perform corrections per-submatrix
<ol style="list-style-type: lower-roman">
<li>orientation correction: i.e. set one feature of interest per axe
which is to capture a uniform pattern for interactions that are towards
upstream or downstream.</li>
<li>ranking pixels per submatrix to highligt the most significant
interaction signals per submatrix.</li>
</ol>
</li>
<li>summarize visually the genome-wide interactions between features of
interest graphically through APA</li>
<li>and/or get individual values of selected zones from each submatrix
for comparison between conditions</li>
</ol>
<p>We also propose (since version 0.99.3) a function to find couples
that demonstrate a non-random (or non-background) like interaction
signals. The function performs z.test to compare target couples to
less-plausible or background couples to identify target couples with
significantly higher interaction values than the background signal
(<code>compareToBackground</code>).</p>
</div>
<div class="section level3">
<h3 id="quickstart-and-tutorials">Quickstart and tutorials:<a class="anchor" aria-label="anchor" href="#quickstart-and-tutorials"></a>
</h3>
<p>Please visit dedicated <a href="https://cuvierlab.github.io/HicAggR/" class="external-link">github pages</a> and the <a href="https://github.com/cuvierlab/HicAggR/" class="external-link">github repository</a>.</p>
</div>
<div class="section level3">
<h3 id="aim">Aim:<a class="anchor" aria-label="anchor" href="#aim"></a>
</h3>
<p>The principal objective is to simplify the extraction of interaction
signals from HiC data, by making submatrices easy to access and treat if
necessary. As such the contact matrices are imported as a list of
<code>ContactMatrix</code> objects per combination of chromosomes(eg:
“chr1_chr1” or “chr1_chr2”). This allows the user to access the data
easily for dowstream analyses.</p>
</div>
</div>
<div class="section level2">
<h2 id="requirements">Requirements<a class="anchor" aria-label="anchor" href="#requirements"></a>
</h2>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>The bioconductor release version:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"HicAggR"</span><span class="op">)</span></span></code></pre></div>
<p>The developmental version:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"CuvierLab/HicAggR"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-library">Load library<a class="anchor" aria-label="anchor" href="#load-library"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/HicAggR" class="external-link">"HicAggR"</a></span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="test-dataset">Test dataset<a class="anchor" aria-label="anchor" href="#test-dataset"></a>
</h2>
<div class="section level3">
<h3 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h3>
<p>Data were obtained from <em>Drosophila melanogaster S2 cells</em>. 1.
HiC test dataset Directly downloaded from the <a href="https://data.4dnucleome.org" class="external-link">4DN</a> platform. * Control
Condition<br>
* Heat Shock Condition<br>
2. Genomic coordinates:<br>
* ChIPseq peaks of Beaf-32 protein in wild type cells (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM1278639" class="external-link">GSM1278639</a>).
* Reference annotation data for TSS from the UCSC database.<br>
* Topologically associating domains (TAD) annotations were defined as
described (<a href="https://doi.org/10.1038/s41467-017-02525-w" class="external-link">F.
Ramirez, 2018</a>).</p>
</div>
<div class="section level3">
<h3 id="genomic-3d-structure">1. Genomic 3D structure<a class="anchor" aria-label="anchor" href="#genomic-3d-structure"></a>
</h3>
<p>For a test, please download HiC data in .hic format (Juicer) and
.mcool format (HiCExplorer). Examples for each format are provided
below.</p>
<div class="section level4">
<h4 id="temp-directory-preparation">Temp directory preparation<a class="anchor" aria-label="anchor" href="#temp-directory-preparation"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_options.html" class="external-link">local_options</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fl">3600</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># temp.dir &lt;- file.path(tempdir(), "HIC_DATA")</span></span>
<span><span class="va">cache.dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu">tools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/tools/userdir.html" class="external-link">R_user_dir</a></span><span class="op">(</span><span class="st">""</span>, which<span class="op">=</span><span class="st">"cache"</span><span class="op">)</span>,<span class="st">".HicAggR_HIC_DATA"</span><span class="op">)</span></span>
<span><span class="va">bfc</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">BiocFileCache</a></span><span class="op">(</span><span class="va">cache.dir</span>,ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co"># dir.create(temp.dir)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="control-condition--hic-file">Control condition (.hic File)<a class="anchor" aria-label="anchor" href="#control-condition--hic-file"></a>
</h4>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcinfo</a></span><span class="op">(</span><span class="va">bfc</span><span class="op">)</span><span class="op">$</span><span class="va">rname</span><span class="op">)</span><span class="op">==</span><span class="fl">0</span> <span class="op">||</span></span>
<span>    <span class="op">!</span><span class="st">"Control_HIC.hic"</span><span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcinfo</a></span><span class="op">(</span><span class="va">bfc</span><span class="op">)</span><span class="op">$</span><span class="va">rname</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">Hic.url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"https://4dn-open-data-public.s3.amazonaws.com/"</span>,</span>
<span>        <span class="st">"fourfront-webprod/wfoutput/7386f953-8da9-47b0-acb2-931cba810544/"</span>,</span>
<span>        <span class="st">"4DNFIOTPSS3L.hic"</span><span class="op">)</span></span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="va">.Platform</span><span class="op">$</span><span class="va">OS.type</span> <span class="op">==</span> <span class="st">"windows"</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="va">HicOutput.pth</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="va">bfc</span>,rname <span class="op">=</span> <span class="st">"Control_HIC.hic"</span>,</span>
<span>            fpath <span class="op">=</span> <span class="va">Hic.url</span>,</span>
<span>            download <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            config <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"auto"</span>,mode<span class="op">=</span><span class="st">"wb"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="kw">else</span><span class="op">{</span></span>
<span>        <span class="va">HicOutput.pth</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="va">bfc</span>, rname <span class="op">=</span> <span class="st">"Control_HIC.hic"</span>,</span>
<span>            fpath <span class="op">=</span> <span class="va">Hic.url</span>,</span>
<span>            download <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            config <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"auto"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span><span class="kw">else</span><span class="op">{</span></span>
<span>    <span class="va">HicOutput.pth</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcpath</a></span><span class="op">(</span><span class="va">bfc</span><span class="op">)</span><span class="op">[</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcinfo</a></span><span class="op">(</span><span class="va">bfc</span><span class="op">)</span><span class="op">$</span><span class="va">rname</span><span class="op">==</span><span class="st">"Control_HIC.hic"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="heat-shock-condition--mcool-file">Heat shock condition (.mcool File)<a class="anchor" aria-label="anchor" href="#heat-shock-condition--mcool-file"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcinfo</a></span><span class="op">(</span><span class="va">bfc</span><span class="op">)</span><span class="op">$</span><span class="va">rname</span><span class="op">)</span><span class="op">==</span><span class="fl">0</span> <span class="op">||</span></span>
<span>    <span class="op">!</span><span class="st">"HeatShock_HIC.mcool"</span><span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcinfo</a></span><span class="op">(</span><span class="va">bfc</span><span class="op">)</span><span class="op">$</span><span class="va">rname</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">Mcool.url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"https://4dn-open-data-public.s3.amazonaws.com/"</span>,</span>
<span>        <span class="st">"fourfront-webprod/wfoutput/4f1479a2-4226-4163-ba99-837f2c8f4ac0/"</span>,</span>
<span>        <span class="st">"4DNFI8DRD739.mcool"</span><span class="op">)</span></span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="va">.Platform</span><span class="op">$</span><span class="va">OS.type</span> <span class="op">==</span> <span class="st">"windows"</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="va">McoolOutput.pth</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="va">bfc</span>, rname <span class="op">=</span> <span class="st">"HeatShock_HIC.mcool"</span>,</span>
<span>            fpath <span class="op">=</span> <span class="va">Mcool.url</span>,</span>
<span>            download <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            config <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"auto"</span>,mode<span class="op">=</span><span class="st">"wb"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="kw">else</span><span class="op">{</span></span>
<span>        <span class="va">McoolOutput.pth</span> <span class="op">&lt;-</span> <span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcadd</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="va">bfc</span>, rname <span class="op">=</span> <span class="st">"HeatShock_HIC.mcool"</span>,</span>
<span>            fpath <span class="op">=</span> <span class="va">Mcool.url</span>,</span>
<span>            download <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            config <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"auto"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span><span class="kw">else</span><span class="op">{</span></span>
<span>    <span class="va">McoolOutput.pth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcpath</a></span><span class="op">(</span><span class="va">bfc</span><span class="op">)</span><span class="op">[</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu">BiocFileCache</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocFileCache/man/BiocFileCache-class.html" class="external-link">bfcinfo</a></span><span class="op">(</span><span class="va">bfc</span><span class="op">)</span><span class="op">$</span><span class="va">rname</span> <span class="op">==</span> <span class="st">"HeatShock_HIC.mcool"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="genomic-location-and-annotation-data">2 Genomic location and annotation data<a class="anchor" aria-label="anchor" href="#genomic-location-and-annotation-data"></a>
</h3>
<p>These kind of data can be imported in R with <a href="https://bioconductor.org/packages/release/bioc/html/rtracklayer.html" class="external-link">rtracklayer</a>
package.</p>
<div class="section level4">
<h4 id="chipseq-peaks-of-beaf-32-protein">ChIPseq peaks of Beaf-32 protein<a class="anchor" aria-label="anchor" href="#chipseq-peaks-of-beaf-32-protein"></a>
</h4>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"Beaf32_Peaks.gnr"</span><span class="op">)</span></span></code></pre></div>
<details><summary>
View
</summary><table class="table">
<thead><tr class="header">
<th align="right">seq</th>
<th align="center">start</th>
<th align="center">end</th>
<th align="center">strand</th>
<th align="center">name</th>
<th align="center">score</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">2L</td>
<td align="center">35594</td>
<td align="center">35725</td>
<td align="center">*</td>
<td align="center">Beaf32_2</td>
<td align="center">76</td>
</tr>
<tr class="even">
<td align="right">2L</td>
<td align="center">47296</td>
<td align="center">47470</td>
<td align="center">*</td>
<td align="center">Beaf32_3</td>
<td align="center">44</td>
</tr>
<tr class="odd">
<td align="right">2L</td>
<td align="center">65770</td>
<td align="center">65971</td>
<td align="center">*</td>
<td align="center">Beaf32_5</td>
<td align="center">520</td>
</tr>
</tbody>
</table></details>
</div>
<div class="section level4">
<h4 id="tss-annontation">TSS annontation<a class="anchor" aria-label="anchor" href="#tss-annontation"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"TSS_Peaks.gnr"</span><span class="op">)</span></span></code></pre></div>
<details><summary>
View
</summary><table class="table">
<thead><tr class="header">
<th align="right">seq</th>
<th align="center">start</th>
<th align="center">end</th>
<th align="center">strand</th>
<th align="center">name</th>
<th align="center">class</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">2L</td>
<td align="center">71757</td>
<td align="center">71757</td>
<td align="center">+</td>
<td align="center">FBgn0031213</td>
<td align="center">active</td>
</tr>
<tr class="even">
<td align="right">2L</td>
<td align="center">76348</td>
<td align="center">76348</td>
<td align="center">+</td>
<td align="center">FBgn0031214</td>
<td align="center">inactive</td>
</tr>
<tr class="odd">
<td align="right">2L</td>
<td align="center">106903</td>
<td align="center">106903</td>
<td align="center">+</td>
<td align="center">FBgn0005278</td>
<td align="center">active</td>
</tr>
</tbody>
</table></details>
</div>
<div class="section level4">
<h4 id="tads-annotation">TADs annotation<a class="anchor" aria-label="anchor" href="#tads-annotation"></a>
</h4>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"TADs_Domains.gnr"</span><span class="op">)</span></span></code></pre></div>
<details><summary>
View
</summary><table class="table">
<thead><tr class="header">
<th align="right">seq</th>
<th align="center">start</th>
<th align="center">end</th>
<th align="center">strand</th>
<th align="center">name</th>
<th align="center">score</th>
<th align="center">class</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">2L</td>
<td align="center">73104</td>
<td align="center">94543</td>
<td align="center">*</td>
<td align="center">Tad_1</td>
<td align="center">3</td>
<td align="center">active</td>
</tr>
<tr class="even">
<td align="right">2L</td>
<td align="center">94544</td>
<td align="center">102930</td>
<td align="center">*</td>
<td align="center">Tad_2</td>
<td align="center">8</td>
<td align="center">active</td>
</tr>
<tr class="odd">
<td align="right">2L</td>
<td align="center">102931</td>
<td align="center">121473</td>
<td align="center">*</td>
<td align="center">Tad_3</td>
<td align="center">8</td>
<td align="center">active</td>
</tr>
</tbody>
</table></details>
</div>
</div>
<div class="section level3">
<h3 id="additional-genome-informations">Additional genome informations<a class="anchor" aria-label="anchor" href="#additional-genome-informations"></a>
</h3>
<p>Required genomic informations used by the functions during the entire
pipeline are a data.frame containing chromosomes names and sized and the
binSize, corresponding to the HiC matrices at the same resolution.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqlengths.num</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'2L'</span><span class="op">=</span><span class="fl">23513712</span>, <span class="st">'2R'</span><span class="op">=</span><span class="fl">25286936</span><span class="op">)</span></span>
<span><span class="va">chromSizes</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    seqnames   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">seqlengths.num</span> <span class="op">)</span>, </span>
<span>    seqlengths <span class="op">=</span> <span class="va">seqlengths.num</span></span>
<span>    <span class="op">)</span></span>
<span><span class="va">binSize</span> <span class="op">&lt;-</span> <span class="fl">5000</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="import-hic">Import HiC<a class="anchor" aria-label="anchor" href="#import-hic"></a>
</h2>
<p>The package supports the import and normalization of HiC data.</p>
<p>NOTE: Since version 0.99.2, the package supports import of balanced
HiC matrices in .hic, .cool/.mcool formats. It also supports the import
of ‘o/e’ matrices in .hic format.</p>
<div class="section level3">
<h3 id="import">Import<a class="anchor" aria-label="anchor" href="#import"></a>
</h3>
<p>HicAggR can import HiC data stored in the main formats: .hic, .cool,
.mcool, .h5 (since version 0.99.2). The pacakage imports by default the
raw counts. Therefore, it is necessary to perform the balancing and
observed/expected correction steps.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HiC_Ctrl.cmx_lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportHiC.html">ImportHiC</a></span><span class="op">(</span></span>
<span>        file    <span class="op">=</span> <span class="va">HicOutput.pth</span>,</span>
<span>        hicResolution     <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>        chrom_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"2L"</span>, <span class="st">"2L"</span>, <span class="st">"2R"</span><span class="op">)</span>,</span>
<span>        chrom_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"2L"</span>, <span class="st">"2R"</span>, <span class="st">"2R"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HiC_HS.cmx_lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ImportHiC.html">ImportHiC</a></span><span class="op">(</span></span>
<span>        file    <span class="op">=</span> <span class="va">McoolOutput.pth</span>,</span>
<span>        hicResolution     <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>        chrom_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"2L"</span>, <span class="st">"2L"</span>, <span class="st">"2R"</span><span class="op">)</span>,</span>
<span>        chrom_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"2L"</span>, <span class="st">"2R"</span>, <span class="st">"2R"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="balancing">Balancing<a class="anchor" aria-label="anchor" href="#balancing"></a>
</h3>
<p>The balancing is done such that every bin of the matrix has
approximately the same number of contacts within the contactMatrix.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HiC_Ctrl.cmx_lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/BalanceHiC.html">BalanceHiC</a></span><span class="op">(</span><span class="va">HiC_Ctrl.cmx_lst</span><span class="op">)</span></span>
<span><span class="va">HiC_HS.cmx_lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/BalanceHiC.html">BalanceHiC</a></span><span class="op">(</span><span class="va">HiC_HS.cmx_lst</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="tips">Tips<a class="anchor" aria-label="anchor" href="#tips"></a>
</h4>
<ol style="list-style-type: decimal">
<li>In the interactionType parameter it is required to define “cis” or
“trans”. Then the function will return only ContactMatrices in the
corresponding category (“cis” or “trans”). All other categories will be
removed from the result.</li>
<li>In the interactionType parameter if you type c(“cis”,“trans”) the
function will normalize separetly “cis” or “trans”. If you type “all”
the function will normalize “cis” and “trans” matrices together.</li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="observedexpected-correction">Observed/Expected Correction<a class="anchor" aria-label="anchor" href="#observedexpected-correction"></a>
</h3>
<p>To correct effects due to genomic distance the matrix is corrected by
the expected values for each genomic distance. The expected values are
by default calculated as the average values of contacts per chromosome
and per distance.</p>
<p>NOTE: Since version 0.99.3, 2 more options to calculate expected
values have been implemented. We designated the methods as the
“lieberman” and the “mean_total”. These methods were implemented based
on the options proposed by HiCExplorer’s <a href="https://hicexplorer.readthedocs.io/en/latest/content/tools/hicTransform.html" class="external-link">hicTransform</a>
program. The “lieberman” method computes per distance (d) expected
values by dividing the sum of all contacts by the difference of
chromosome length and distance(d).</p>
<p>The “mean_total” is simply the average of all contact values
including 0 values, which are ignored in the default method
(“mean_non_zero”)</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HiC_Ctrl.cmx_lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/OverExpectedHiC.html">OverExpectedHiC</a></span><span class="op">(</span><span class="va">HiC_Ctrl.cmx_lst</span><span class="op">)</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HiC_HS.cmx_lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/OverExpectedHiC.html">OverExpectedHiC</a></span><span class="op">(</span><span class="va">HiC_HS.cmx_lst</span><span class="op">)</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-20-2.png" width="700"></p>
<div class="section level4">
<h4 id="tips-1">Tips<a class="anchor" aria-label="anchor" href="#tips-1"></a>
</h4>
<ol style="list-style-type: decimal">
<li>After runing <code>OverExpectedHiC</code> function, expected counts
can be plotted as a function of genomic distances with tibble by taking
the expected attributes.</li>
</ol>
</div>
</div>
<div class="section level3">
<h3 id="hic-data-format-contactmatrix-list-structure">HiC data format: ContactMatrix list structure<a class="anchor" aria-label="anchor" href="#hic-data-format-contactmatrix-list-structure"></a>
</h3>
<p>Each element of the list corresponds to a ContactMatrix object
(dgCMatrix object, sparse matrix format) storing contact frequencies for
one chromosome (cis-interactions, ex: “2L_2L”) or between two
chromosomes (trans-interactions, ex: “2L_2R”). HiC data format is based
on <a href="https://bioconductor.org/packages/release/bioc/html/InteractionSet.html" class="external-link">InteractionSet</a>
and <a href="https://cran.r-project.org/web/packages/Matrix/Matrix.pdf" class="external-link">Matrix</a>
packages.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">HiC_Ctrl.cmx_lst</span>,max.level <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 3</span></span>
<span><span class="co">#&gt;  $ 2L_2L:Formal class 'ContactMatrix' [package "InteractionSet"] with 5 slots</span></span>
<span><span class="co">#&gt;   .. ..@ matrix  :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. ..@ anchor1 : int [1:4703] 1 2 3 4 5 6 7 8 9 10 ...</span></span>
<span><span class="co">#&gt;   .. ..@ anchor2 : int [1:4703] 1 2 3 4 5 6 7 8 9 10 ...</span></span>
<span><span class="co">#&gt;   .. ..@ regions :Formal class 'GRanges' [package "GenomicRanges"] with 7 slots</span></span>
<span><span class="co">#&gt;   .. ..@ metadata:List of 10</span></span>
<span><span class="co">#&gt;   .. .. ..$ name         : chr "2L_2L"</span></span>
<span><span class="co">#&gt;   .. .. ..$ type         : chr "cis"</span></span>
<span><span class="co">#&gt;   .. .. ..$ kind         : chr "U"</span></span>
<span><span class="co">#&gt;   .. .. ..$ symmetric    : logi TRUE</span></span>
<span><span class="co">#&gt;   .. .. ..$ resolution   : num 5000</span></span>
<span><span class="co">#&gt;   .. .. ..$ expected     : num [1:2410023] 719 673 719 370 673 ...</span></span>
<span><span class="co">#&gt;   .. .. ..$ removedCounts:Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. .. ..$ observed     : num [1:2410023] 1329 1721 1852 1076 2191 ...</span></span>
<span><span class="co">#&gt;   .. .. ..$ normalizer   : num [1:2410023] 1.399 1.115 0.888 1.139 0.908 ...</span></span>
<span><span class="co">#&gt;   .. .. ..$ mtx          : chr "norm"</span></span>
<span><span class="co">#&gt;  $ 2L_2R:Formal class 'ContactMatrix' [package "InteractionSet"] with 5 slots</span></span>
<span><span class="co">#&gt;   .. ..@ matrix  :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. ..@ anchor1 : int [1:4703] 1 2 3 4 5 6 7 8 9 10 ...</span></span>
<span><span class="co">#&gt;   .. ..@ anchor2 : int [1:5058] 4704 4705 4706 4707 4708 4709 4710 4711 4712 4713 ...</span></span>
<span><span class="co">#&gt;   .. ..@ regions :Formal class 'GRanges' [package "GenomicRanges"] with 7 slots</span></span>
<span><span class="co">#&gt;   .. ..@ metadata:List of 9</span></span>
<span><span class="co">#&gt;   .. .. ..$ name         : chr "2L_2R"</span></span>
<span><span class="co">#&gt;   .. .. ..$ type         : chr "trans"</span></span>
<span><span class="co">#&gt;   .. .. ..$ kind         : chr NA</span></span>
<span><span class="co">#&gt;   .. .. ..$ symmetric    : logi FALSE</span></span>
<span><span class="co">#&gt;   .. .. ..$ resolution   : num 5000</span></span>
<span><span class="co">#&gt;   .. .. ..$ removedCounts:Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. .. ..$ observed     : num [1:1292193] 1 1 1 1 1 1 1 1 1 1 ...</span></span>
<span><span class="co">#&gt;   .. .. ..$ normalizer   : num [1:1292193] 2.73 2.82 2.78 2.48 4.4 ...</span></span>
<span><span class="co">#&gt;   .. .. ..$ expected     : num 0.0596</span></span>
<span><span class="co">#&gt;  $ 2R_2R:Formal class 'ContactMatrix' [package "InteractionSet"] with 5 slots</span></span>
<span><span class="co">#&gt;   .. ..@ matrix  :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. ..@ anchor1 : int [1:5058] 1 2 3 4 5 6 7 8 9 10 ...</span></span>
<span><span class="co">#&gt;   .. ..@ anchor2 : int [1:5058] 1 2 3 4 5 6 7 8 9 10 ...</span></span>
<span><span class="co">#&gt;   .. ..@ regions :Formal class 'GRanges' [package "GenomicRanges"] with 7 slots</span></span>
<span><span class="co">#&gt;   .. ..@ metadata:List of 10</span></span>
<span><span class="co">#&gt;   .. .. ..$ name         : chr "2R_2R"</span></span>
<span><span class="co">#&gt;   .. .. ..$ type         : chr "cis"</span></span>
<span><span class="co">#&gt;   .. .. ..$ kind         : chr "U"</span></span>
<span><span class="co">#&gt;   .. .. ..$ symmetric    : logi TRUE</span></span>
<span><span class="co">#&gt;   .. .. ..$ resolution   : num 5000</span></span>
<span><span class="co">#&gt;   .. .. ..$ expected     : num [1:2769177] 746 703 746 152 192 ...</span></span>
<span><span class="co">#&gt;   .. .. ..$ removedCounts:Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. .. ..$ observed     : num [1:2769177] 374 368 614 95 131 420 18 17 26 332 ...</span></span>
<span><span class="co">#&gt;   .. .. ..$ normalizer   : num [1:2769177] 2.62 2.19 1.84 2.94 2.46 ...</span></span>
<span><span class="co">#&gt;   .. .. ..$ mtx          : chr "norm"</span></span>
<span><span class="co">#&gt;  - attr(*, "resolution")= num 5000</span></span>
<span><span class="co">#&gt;  - attr(*, "chromSize")= tibble [2 × 3] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   ..$ name     : chr [1:2] "2L" "2R"</span></span>
<span><span class="co">#&gt;   ..$ length   : num [1:2] 23513712 25286936</span></span>
<span><span class="co">#&gt;   ..$ dimension: num [1:2] 4703 5058</span></span>
<span><span class="co">#&gt;  - attr(*, "matricesKind")= tibble [3 × 4] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   ..$ name     : chr [1:3] "2L_2L" "2L_2R" "2R_2R"</span></span>
<span><span class="co">#&gt;   ..$ type     : chr [1:3] "cis" "trans" "cis"</span></span>
<span><span class="co">#&gt;   ..$ kind     : chr [1:3] "U" NA "U"</span></span>
<span><span class="co">#&gt;   ..$ symmetric: logi [1:3] TRUE FALSE TRUE</span></span>
<span><span class="co">#&gt;  - attr(*, "mtx")= chr "o/e"</span></span>
<span><span class="co">#&gt;  - attr(*, "expected")= tibble [4,961 × 2] (S3: tbl_df/tbl/data.frame)</span></span>
<span><span class="co">#&gt;   ..$ distance: num [1:4961] 1 5001 10001 15001 20001 ...</span></span>
<span><span class="co">#&gt;   ..$ expected: num [1:4961] 733 689 377 257 191 ...</span></span>
<span><span class="co">#&gt;</span></span></code></pre></div>
<p>The list has the attributes described below. These attributes are
accessible via:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">HiC_Ctrl.cmx_lst</span><span class="op">)</span></span>
<span><span class="co">#&gt; $names</span></span>
<span><span class="co">#&gt; [1] "2L_2L" "2L_2R" "2R_2R"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $resolution</span></span>
<span><span class="co">#&gt; [1] 5000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $chromSize</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 3</span></span></span>
<span><span class="co">#&gt;   name    length dimension</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 2L    23<span style="text-decoration: underline;">513</span>712      <span style="text-decoration: underline;">4</span>703</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 2R    25<span style="text-decoration: underline;">286</span>936      <span style="text-decoration: underline;">5</span>058</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $matricesKind</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 4</span></span></span>
<span><span class="co">#&gt;   name  type  kind  symmetric</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 2L_2L cis   U     TRUE     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 2L_2R trans <span style="color: #BB0000;">NA</span>    FALSE    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 2R_2R cis   U     TRUE     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $mtx</span></span>
<span><span class="co">#&gt; [1] "o/e"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $expected</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4,961 × 2</span></span></span>
<span><span class="co">#&gt;   distance expected</span></span>
<span><span class="co">#&gt;      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>        1     733.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     <span style="text-decoration: underline;">5</span>001     689.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>    <span style="text-decoration: underline;">10</span>001     377.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>    <span style="text-decoration: underline;">15</span>001     257.</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 4,957 more rows</span></span></span>
<span><span class="co">#&gt;</span></span></code></pre></div>
<ol style="list-style-type: decimal">
<li>
<strong>names</strong> : the names of list elements
(ContactMatrix).</li>
<li>
<strong>resolution</strong> : the resolution of the HiC map.<br>
</li>
<li>
<strong>chromSize</strong> : the size of the chromosomes in the
tibble format.<br>
- <em>seqnames</em> : the sequence name (chromosome name).<br>
- <em>seqlengths</em> : the sequence length in base pairs.<br>
- <em>dimension</em> : the sequence length in number of bins.</li>
<li>
<strong>matricesKind</strong> : the kind of matrix that composes the
list in the tibble format.<br>
- <em>name</em> : the matrix name. - <em>type</em> : interactionType.
<em>“Cis”</em> for interactions on the same chromosome and
<em>“Trans”</em> for interactions on different chromosomes. -
<em>kind</em> : the matrix kind. U for upper triangle matrices, L for
lower triangle matrices, NA for rectangular or square matrices. -
<em>symmetric</em> : a boolean that indicates whether the matrix is
symmetric (lower triangle identical to upper triangle).<br>
</li>
<li>
<strong>mtx</strong> : the kind of values in matrix. For exemple
observed counts, normalized counts, observed/expected, etc.</li>
<li>
<strong>expected</strong> : This attribute is related to the
<code>OverExpectedHiC</code> function. It gives a tibble with the
expected counts as a function of genomic distance.</li>
</ol>
<p>Each contactmatrix in the list have metadata. These are accessible
via:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">HiC_Ctrl.cmx_lst</span><span class="op">[[</span><span class="st">"2L_2L"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 10</span></span>
<span><span class="co">#&gt;  $ name         : chr "2L_2L"</span></span>
<span><span class="co">#&gt;  $ type         : chr "cis"</span></span>
<span><span class="co">#&gt;  $ kind         : chr "U"</span></span>
<span><span class="co">#&gt;  $ symmetric    : logi TRUE</span></span>
<span><span class="co">#&gt;  $ resolution   : num 5000</span></span>
<span><span class="co">#&gt;  $ expected     : num [1:2410023] 719 673 719 370 673 ...</span></span>
<span><span class="co">#&gt;  $ removedCounts:Formal class 'dgCMatrix' [package "Matrix"] with 6 slots</span></span>
<span><span class="co">#&gt;   .. ..@ i       : int [1:266585] 2 10 18 19 20 28 32 37 38 40 ...</span></span>
<span><span class="co">#&gt;   .. ..@ p       : int [1:4704] 0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">#&gt;   .. ..@ Dim     : int [1:2] 4703 4703</span></span>
<span><span class="co">#&gt;   .. ..@ Dimnames:List of 2</span></span>
<span><span class="co">#&gt;   .. .. ..$ : NULL</span></span>
<span><span class="co">#&gt;   .. .. ..$ : NULL</span></span>
<span><span class="co">#&gt;   .. ..@ x       : num [1:266585] 1 1 1 2 1 1 1 1 1 2 ...</span></span>
<span><span class="co">#&gt;   .. ..@ factors : list()</span></span>
<span><span class="co">#&gt;  $ observed     : num [1:2410023] 1329 1721 1852 1076 2191 ...</span></span>
<span><span class="co">#&gt;  $ normalizer   : num [1:2410023] 1.399 1.115 0.888 1.139 0.908 ...</span></span>
<span><span class="co">#&gt;  $ mtx          : chr "norm"</span></span>
<span><span class="co">#&gt;</span></span></code></pre></div>
<ol style="list-style-type: decimal">
<li>
<strong>name</strong> : the name of the ContactMatrix.</li>
<li>
<strong>type</strong> : interactionType. <em>“Cis”</em> for
interactions on the same chromosome and <em>“Trans”</em> for
interactions on different chromosomes (or arms).<br>
</li>
<li>
<strong>kind</strong> : the matrix kind. U for upper triangle
matrices, L for lower triangle matrices, NA for rectangular or
square</li>
<li>
<strong>symmetric</strong> : a boolean that indicates whether the
matrix is symmetric (lower triangle identical to upper triangle).<br>
</li>
<li>
<strong>resolution</strong> : resolution of the HiC map.</li>
<li>
<strong>removedCounts</strong> : A sparse matrix (dgCMatrix) of the
removed counts (counts that are below the threshold on rows or columns
as described in <code>BalanceHiC</code>).</li>
<li>
<strong>observed</strong> : observed counts of the sparse
matrix.</li>
<li>
<strong>normalizer</strong> : the balancer vector that converts the
observed counts into normalized counts. (observed * normalizer =
normalized counts).<br>
</li>
<li>
<strong>mtx</strong> : the kind of values in matrix. For example
observed counts, normalized counts, observed/expected, etc.</li>
<li>
<strong>expected</strong> : This attributes is related to the
<code>OverExpectedHiC</code> function. It gives the expected vector that
convert the normalized counts into the observed/expected counts
(normalized counts / expected = observed/expected).</li>
</ol>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="indexing">Indexing<a class="anchor" aria-label="anchor" href="#indexing"></a>
</h2>
<p>This part of the data corresponds to the positioning data (ChIPseq
peaks, genomic features and annotations, genes, etc) on the genome. To
integrate such annotations with HiC data in 2D matrices, annotations
must be processed as followed.</p>
<p>The first step is the indexing of the features. It allows the
features to be splitted and grouped into bins corresponding to the HiC
bin size.</p>
<div class="section level3">
<h3 id="example-1-anchors-from-beaf32-chip-seq-peaks-bed-file">Example 1: Anchors from Beaf32 ChIP-seq peaks (bed file)<a class="anchor" aria-label="anchor" href="#example-1-anchors-from-beaf32-chip-seq-peaks-bed-file"></a>
</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anchors_Index.gnr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IndexFeatures.html">IndexFeatures</a></span><span class="op">(</span></span>
<span>    gRangeList        <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Beaf<span class="op">=</span><span class="va">Beaf32_Peaks.gnr</span><span class="op">)</span>, </span>
<span>    genomicConstraint        <span class="op">=</span> <span class="va">TADs_Domains.gnr</span>,</span>
<span>    chromSizes         <span class="op">=</span> <span class="va">chromSizes</span>,</span>
<span>    binSize           <span class="op">=</span> <span class="va">binSize</span>,</span>
<span>    metadataColName <span class="op">=</span> <span class="st">"score"</span>,</span>
<span>    method            <span class="op">=</span> <span class="st">"max"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<details><summary>
View
</summary><table class="table">
<colgroup>
<col width="9%">
<col width="7%">
<col width="7%">
<col width="6%">
<col width="7%">
<col width="12%">
<col width="6%">
<col width="11%">
<col width="11%">
<col width="10%">
<col width="9%">
</colgroup>
<thead><tr class="header">
<th align="left">seqnames</th>
<th align="right">start</th>
<th align="right">end</th>
<th align="right">width</th>
<th align="left">strand</th>
<th align="left">name</th>
<th align="left">bin</th>
<th align="left">constraint</th>
<th align="right">Beaf.score</th>
<th align="left">Beaf.name</th>
<th align="left">Beaf.bln</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">2L</td>
<td align="right">70001</td>
<td align="right">75000</td>
<td align="right">5000</td>
<td align="left">*</td>
<td align="left">2L:15:Tad_1</td>
<td align="left">2L:15</td>
<td align="left">Tad_1</td>
<td align="right">205</td>
<td align="left">Beaf32_8</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">2L</td>
<td align="right">100001</td>
<td align="right">105000</td>
<td align="right">5000</td>
<td align="left">*</td>
<td align="left">2L:21:Tad_2</td>
<td align="left">2L:21</td>
<td align="left">Tad_2</td>
<td align="right">1830</td>
<td align="left">Beaf32_11</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">2L</td>
<td align="right">110001</td>
<td align="right">115000</td>
<td align="right">5000</td>
<td align="left">*</td>
<td align="left">2L:23:Tad_3</td>
<td align="left">2L:23</td>
<td align="left">Tad_3</td>
<td align="right">1707</td>
<td align="left">Beaf32_14</td>
<td align="left">TRUE</td>
</tr>
</tbody>
</table></details>
</div>
<div class="section level3">
<h3 id="example-2-baits-from-tss-transcription-start-sites-from-ucsc">Example 2: Baits from TSS (transcription start sites from UCSC)<a class="anchor" aria-label="anchor" href="#example-2-baits-from-tss-transcription-start-sites-from-ucsc"></a>
</h3>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">baits_Index.gnr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IndexFeatures.html">IndexFeatures</a></span><span class="op">(</span></span>
<span>    gRangeList        <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Tss<span class="op">=</span><span class="va">TSS_Peaks.gnr</span><span class="op">)</span>,</span>
<span>    genomicConstraint        <span class="op">=</span> <span class="va">TADs_Domains.gnr</span>,</span>
<span>    chromSizes         <span class="op">=</span> <span class="va">chromSizes</span>,</span>
<span>    binSize           <span class="op">=</span> <span class="va">binSize</span>,</span>
<span>    metadataColName <span class="op">=</span> <span class="st">"score"</span>,</span>
<span>    method            <span class="op">=</span> <span class="st">"max"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<details><summary>
View
</summary><table class="table">
<colgroup>
<col width="9%">
<col width="7%">
<col width="7%">
<col width="6%">
<col width="7%">
<col width="12%">
<col width="6%">
<col width="11%">
<col width="10%">
<col width="13%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="left">seqnames</th>
<th align="right">start</th>
<th align="right">end</th>
<th align="right">width</th>
<th align="left">strand</th>
<th align="left">name</th>
<th align="left">bin</th>
<th align="left">constraint</th>
<th align="left">Tss.class</th>
<th align="left">Tss.name</th>
<th align="left">Tss.bln</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">2L</td>
<td align="right">75001</td>
<td align="right">80000</td>
<td align="right">5000</td>
<td align="left">*</td>
<td align="left">2L:16:Tad_1</td>
<td align="left">2L:16</td>
<td align="left">Tad_1</td>
<td align="left">inactive</td>
<td align="left">FBgn0031214</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">2L</td>
<td align="right">105001</td>
<td align="right">110000</td>
<td align="right">5000</td>
<td align="left">*</td>
<td align="left">2L:22:Tad_3</td>
<td align="left">2L:22</td>
<td align="left">Tad_3</td>
<td align="left">active</td>
<td align="left">FBgn0026….</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">2L</td>
<td align="right">115001</td>
<td align="right">120000</td>
<td align="right">5000</td>
<td align="left">*</td>
<td align="left">2L:24:Tad_3</td>
<td align="left">2L:24</td>
<td align="left">Tad_3</td>
<td align="left">active</td>
<td align="left">FBgn0031219</td>
<td align="left">TRUE</td>
</tr>
</tbody>
</table></details>
</div>
<div class="section level3">
<h3 id="filter-indexed-features">Filter indexed features:<a class="anchor" aria-label="anchor" href="#filter-indexed-features"></a>
</h3>
<p>By using features names and bin IDs, it is possible to filter a
subset of features. Example: Subset TSS that are not in the same bin
than a Beaf32 peak.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">non_Overlaps.ndx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">baits_Index.gnr</span><span class="op">$</span><span class="va">bin</span>, </span>
<span>    <span class="va">anchors_Index.gnr</span><span class="op">$</span><span class="va">bin</span>, nomatch<span class="op">=</span><span class="fl">0L</span><span class="op">)</span><span class="op">==</span><span class="fl">0L</span></span>
<span><span class="va">baits_Index.gnr</span> <span class="op">&lt;-</span> <span class="va">baits_Index.gnr</span><span class="op">[</span><span class="va">non_Overlaps.ndx</span>,<span class="op">]</span></span></code></pre></div>
<details><summary>
View
</summary><table class="table">
<colgroup>
<col width="9%">
<col width="7%">
<col width="7%">
<col width="6%">
<col width="7%">
<col width="12%">
<col width="6%">
<col width="11%">
<col width="10%">
<col width="13%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="left">seqnames</th>
<th align="right">start</th>
<th align="right">end</th>
<th align="right">width</th>
<th align="left">strand</th>
<th align="left">name</th>
<th align="left">bin</th>
<th align="left">constraint</th>
<th align="left">Tss.class</th>
<th align="left">Tss.name</th>
<th align="left">Tss.bln</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">2L</td>
<td align="right">75001</td>
<td align="right">80000</td>
<td align="right">5000</td>
<td align="left">*</td>
<td align="left">2L:16:Tad_1</td>
<td align="left">2L:16</td>
<td align="left">Tad_1</td>
<td align="left">inactive</td>
<td align="left">FBgn0031214</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">2L</td>
<td align="right">105001</td>
<td align="right">110000</td>
<td align="right">5000</td>
<td align="left">*</td>
<td align="left">2L:22:Tad_3</td>
<td align="left">2L:22</td>
<td align="left">Tad_3</td>
<td align="left">active</td>
<td align="left">FBgn0026….</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">2L</td>
<td align="right">115001</td>
<td align="right">120000</td>
<td align="right">5000</td>
<td align="left">*</td>
<td align="left">2L:24:Tad_3</td>
<td align="left">2L:24</td>
<td align="left">Tad_3</td>
<td align="left">active</td>
<td align="left">FBgn0031219</td>
<td align="left">TRUE</td>
</tr>
</tbody>
</table></details>
</div>
<div class="section level3">
<h3 id="tips-2">Tips<a class="anchor" aria-label="anchor" href="#tips-2"></a>
</h3>
<ol style="list-style-type: decimal">
<li>It is possible to index multiple features at the same time by
submitting a named list of GRanges. Names given in the list of GRanges
can then be used to filter indexed features and pairs.<br>
</li>
<li>If genomicConstraint is defined, then anchors and baits will be
paired when located within the same region only. If contraint.gnr is
NULL, entire chromosomes (or arms) are used as constraints.<br>
</li>
<li>When multiple ranges are in a same bin (ex: 3 ChIP-seq peaks in the
same 10kb bin), associated numeric variables in metadata
(<code>metadataColName</code>) can be summarized according to the
defined method (<code>method</code>), Example: Max peak score of the bin
is kept in metadata column <code>score</code>.</li>
</ol>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="search-pairs">Search Pairs<a class="anchor" aria-label="anchor" href="#search-pairs"></a>
</h2>
<div class="section level3">
<h3 id="pairing">Pairing<a class="anchor" aria-label="anchor" href="#pairing"></a>
</h3>
<p><code>SearchPairs</code> function takes as input one or two indexed
features and returns all putative pairs within the same constraint (ex:
wihtin the same TAD).<br>
If only one indexed features is defined in indexAnchor,
<code>SearchPairs</code> will return symetrical homotypic pairs
(A&lt;-&gt;A), if indexAnchor and indexBait are defined, it will return
asymetrical heterotypic pairs (A&lt;-&gt;B).</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">interactions.gni</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SearchPairs.html">SearchPairs</a></span><span class="op">(</span></span>
<span>        indexAnchor <span class="op">=</span> <span class="va">anchors_Index.gnr</span>,</span>
<span>        indexBait   <span class="op">=</span> <span class="va">baits_Index.gnr</span></span>
<span>        <span class="op">)</span></span></code></pre></div>
<details><summary>
View
</summary><table class="table">
<thead>
<tr>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="6">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
Ranges
</div>
</th>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="15">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
Metadata
</div>
</th>
</tr>
<tr>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="3">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
First
</div>
</th>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="3">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
Second
</div>
</th>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="5">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
Interaction
</div>
</th>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="5">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
Anchor
</div>
</th>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="5">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
Bait
</div>
</th>
</tr>
<tr>
<th style="text-align:right;">
seq
</th>
<th style="text-align:center;">
start
</th>
<th style="text-align:center;">
end
</th>
<th style="text-align:right;">
seq
</th>
<th style="text-align:center;">
start
</th>
<th style="text-align:center;">
end
</th>
<th style="text-align:center;">
name
</th>
<th style="text-align:center;">
constraint
</th>
<th style="text-align:center;">
distance
</th>
<th style="text-align:center;">
orientation
</th>
<th style="text-align:center;">
submatrix.name
</th>
<th style="text-align:center;">
name
</th>
<th style="text-align:center;">
bin
</th>
<th style="text-align:center;">
Beaf.name
</th>
<th style="text-align:center;">
Beaf.score
</th>
<th style="text-align:center;">
Beaf.bln
</th>
<th style="text-align:center;">
name
</th>
<th style="text-align:center;">
bin
</th>
<th style="text-align:center;">
Tss.name
</th>
<th style="text-align:center;">
Tss.class
</th>
<th style="text-align:center;">
Tss.bln
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
70001
</td>
<td style="text-align:center;">
75000
</td>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
75001
</td>
<td style="text-align:center;">
80000
</td>
<td style="text-align:center;">
2L:15_2L:16
</td>
<td style="text-align:center;">
Tad_1
</td>
<td style="text-align:center;">
5000
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
2L:15_2L:16
</td>
<td style="text-align:center;">
2L:15:Tad_1
</td>
<td style="text-align:center;">
2L:15
</td>
<td style="text-align:center;">
Beaf32_8
</td>
<td style="text-align:center;">
205
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
2L:16:Tad_1
</td>
<td style="text-align:center;">
2L:16
</td>
<td style="text-align:center;">
FBgn0031214
</td>
<td style="text-align:center;">
inactive
</td>
<td style="text-align:center;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
110001
</td>
<td style="text-align:center;">
115000
</td>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
105001
</td>
<td style="text-align:center;">
110000
</td>
<td style="text-align:center;">
2L:23_2L:22
</td>
<td style="text-align:center;">
Tad_3
</td>
<td style="text-align:center;">
5000
</td>
<td style="text-align:center;">
FALSE
</td>
<td style="text-align:center;">
2L:22_2L:23
</td>
<td style="text-align:center;">
2L:23:Tad_3
</td>
<td style="text-align:center;">
2L:23
</td>
<td style="text-align:center;">
Beaf32_14
</td>
<td style="text-align:center;">
1707
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
2L:22:Tad_3
</td>
<td style="text-align:center;">
2L:22
</td>
<td style="text-align:center;">
FBgn0026….
</td>
<td style="text-align:center;">
active
</td>
<td style="text-align:center;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
120001
</td>
<td style="text-align:center;">
125000
</td>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
105001
</td>
<td style="text-align:center;">
110000
</td>
<td style="text-align:center;">
2L:25_2L:22
</td>
<td style="text-align:center;">
Tad_3
</td>
<td style="text-align:center;">
15000
</td>
<td style="text-align:center;">
FALSE
</td>
<td style="text-align:center;">
2L:22_2L:25
</td>
<td style="text-align:center;">
2L:25:Tad_3
</td>
<td style="text-align:center;">
2L:25
</td>
<td style="text-align:center;">
Beaf32_15
</td>
<td style="text-align:center;">
484
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
2L:22:Tad_3
</td>
<td style="text-align:center;">
2L:22
</td>
<td style="text-align:center;">
FBgn0026….
</td>
<td style="text-align:center;">
active
</td>
<td style="text-align:center;">
TRUE
</td>
</tr>
</tbody>
</table></details>
</div>
<div class="section level3">
<h3 id="tips-3">Tips<a class="anchor" aria-label="anchor" href="#tips-3"></a>
</h3>
<ol style="list-style-type: decimal">
<li>If <code>indexBait</code> is NULL, <code>SearchPairs</code> will
return homotypic pairs with <code>indexAnchor</code>.</li>
<li>Minimum and maximum distances between pairs anchors can be set. Note
that it is also possible to filter pairs within a specific distance
later on.</li>
</ol>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="extractions">Extractions<a class="anchor" aria-label="anchor" href="#extractions"></a>
</h2>
<div class="section level3">
<h3 id="case-1-long-range-interactions-between-two-distal-anchors-">Case 1: Long-range interactions between two distal anchors.<a class="anchor" aria-label="anchor" href="#case-1-long-range-interactions-between-two-distal-anchors-"></a>
</h3>
<p><img src="images/Extractions_of_LRI.png"><!-- --></p>
<div class="section level4">
<h4 id="interactions-defined-with-ginteraction-or-pairs-of-granges-">Interactions defined with GInteraction or Pairs of GRanges.<a class="anchor" aria-label="anchor" href="#interactions-defined-with-ginteraction-or-pairs-of-granges-"></a>
</h4>
<p>In extracted matrices, the middle of the Y axis corresponds to the
center of the first element and interacts with the center of second
element in the middle of the X axis.</p>
<p><img src="images/LRI_GInteractions.png"><!-- --></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">interactions_PFmatrix.lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractSubmatrix.html">ExtractSubmatrix</a></span><span class="op">(</span></span>
<span>    genomicFeature         <span class="op">=</span> <span class="va">interactions.gni</span>,</span>
<span>    hicLst        <span class="op">=</span> <span class="va">HiC_Ctrl.cmx_lst</span>,</span>
<span>    referencePoint <span class="op">=</span> <span class="st">"pf"</span>,</span>
<span>    matriceDim     <span class="op">=</span> <span class="fl">41</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="interactions-defined-with-granges-">Interactions defined with GRanges.<a class="anchor" aria-label="anchor" href="#interactions-defined-with-granges-"></a>
</h4>
<p>The middle of the Y axis corresponds to the start of the range and
interacts with the middle of the X axis which corresponds to the end of
the range.</p>
<p><img src="images/LRI_GRanges.png"><!-- --></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">domains_PFmatrix.lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractSubmatrix.html">ExtractSubmatrix</a></span><span class="op">(</span></span>
<span>    genomicFeature         <span class="op">=</span> <span class="va">TADs_Domains.gnr</span>,</span>
<span>    hicLst        <span class="op">=</span> <span class="va">HiC_Ctrl.cmx_lst</span>,</span>
<span>    referencePoint <span class="op">=</span> <span class="st">"pf"</span>,</span>
<span>    matriceDim     <span class="op">=</span> <span class="fl">41</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="case-2-interactions-around-genomic-regions-or-domains-">Case 2: Interactions around genomic regions or domains.<a class="anchor" aria-label="anchor" href="#case-2-interactions-around-genomic-regions-or-domains-"></a>
</h3>
<p>In this case, extracted matrices are resized and scaled in order to
fit all regions into the same area.</p>
<p><img src="images/Extractions_of_Regions.png"><!-- --></p>
<div class="section level4">
<h4 id="regions-defined-with-ginteraction-object-or-pairs-of-granges">Regions defined with GInteraction object or Pairs of GRanges<a class="anchor" aria-label="anchor" href="#regions-defined-with-ginteraction-object-or-pairs-of-granges"></a>
</h4>
<p>The region’s start is defined by the center of the first element and
the region’s end by the center of the second element. <img src="images/Regions_GInteractions.png"><!-- --></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">interactions_RFmatrix_ctrl.lst</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractSubmatrix.html">ExtractSubmatrix</a></span><span class="op">(</span></span>
<span>    genomicFeature         <span class="op">=</span> <span class="va">interactions.gni</span>,</span>
<span>    hicLst        <span class="op">=</span> <span class="va">HiC_Ctrl.cmx_lst</span>,</span>
<span>    hicResolution            <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    referencePoint <span class="op">=</span> <span class="st">"rf"</span>,</span>
<span>    matriceDim     <span class="op">=</span> <span class="fl">101</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="regions-defined-with-granges">Regions defined with GRanges<a class="anchor" aria-label="anchor" href="#regions-defined-with-granges"></a>
</h4>
<p>The regions are directly defined by the ranges of GRanges object.</p>
<p><img src="images/Regions_GRanges.png"><!-- --></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">domains_RFmatrix.lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractSubmatrix.html">ExtractSubmatrix</a></span><span class="op">(</span></span>
<span>    genomicFeature         <span class="op">=</span> <span class="va">TADs_Domains.gnr</span>,</span>
<span>    hicLst        <span class="op">=</span> <span class="va">HiC_Ctrl.cmx_lst</span>,</span>
<span>    referencePoint <span class="op">=</span> <span class="st">"rf"</span>,</span>
<span>    matriceDim     <span class="op">=</span> <span class="fl">101</span>,</span>
<span>    cores          <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    verbose        <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="case-3-interactions-along-the-chromosome-axis-">Case 3: Interactions along the chromosome axis.<a class="anchor" aria-label="anchor" href="#case-3-interactions-along-the-chromosome-axis-"></a>
</h3>
<p><img src="images/Extractions_of_Ponctuals_Interactions.png"><!-- --></p>
<div class="section level4">
<h4 id="example-to-analyse-interactions-in-the-context-of-tads">Example to analyse interactions in the context of TADs:<a class="anchor" aria-label="anchor" href="#example-to-analyse-interactions-in-the-context-of-tads"></a>
</h4>
<p><strong>Step 1:</strong> generate a GRanges object of TAD boundaries
by concatenating starts and ends of TADs.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">domains_Border.gnr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">resize</a></span><span class="op">(</span><span class="va">TADs_Domains.gnr</span>, <span class="fl">1</span>, <span class="st">"start"</span><span class="op">)</span>,</span>
<span>        <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/intra-range-methods.html" class="external-link">resize</a></span><span class="op">(</span><span class="va">TADs_Domains.gnr</span>, <span class="fl">1</span>,  <span class="st">"end"</span> <span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>

<p><strong>Step 2:</strong> Filter and reduce TAD boundaries GRanges
object according to HiC resolution (binSize) + Store TAD names.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">domains_Border_Bin.gnr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/BinGRanges.html">BinGRanges</a></span><span class="op">(</span></span>
<span>    gRange  <span class="op">=</span> <span class="va">domains_Border.gnr</span>,</span>
<span>    binSize <span class="op">=</span> <span class="va">binSize</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span></span>
<span><span class="va">domains_Border_Bin.gnr</span><span class="op">$</span><span class="va">subname</span> <span class="op">&lt;-</span> <span class="va">domains_Border_Bin.gnr</span><span class="op">$</span><span class="va">name</span></span>
<span><span class="va">domains_Border_Bin.gnr</span><span class="op">$</span><span class="va">name</span>    <span class="op">&lt;-</span> <span class="va">domains_Border_Bin.gnr</span><span class="op">$</span><span class="va">bin</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">domains_Border_Bin.gnr</span></span></code></pre></div>
<details><summary>
View
</summary><table style="width:100%;" class="table">
<colgroup>
<col width="5%">
<col width="11%">
<col width="11%">
<col width="11%">
<col width="9%">
<col width="9%">
<col width="11%">
<col width="9%">
<col width="19%">
</colgroup>
<thead><tr class="header">
<th align="right">seq</th>
<th align="center">start</th>
<th align="center">end</th>
<th align="center">strand</th>
<th align="center">name</th>
<th align="center">score</th>
<th align="center">class</th>
<th align="center">bin</th>
<th align="center">subname</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">2L</td>
<td align="center">70001</td>
<td align="center">75000</td>
<td align="center">*</td>
<td align="center">2L:15</td>
<td align="center">3</td>
<td align="center">active</td>
<td align="center">2L:15</td>
<td align="center">Tad_1</td>
</tr>
<tr class="even">
<td align="right">2L</td>
<td align="center">90001</td>
<td align="center">95000</td>
<td align="center">*</td>
<td align="center">2L:19</td>
<td align="center">3, 8</td>
<td align="center">active</td>
<td align="center">2L:19</td>
<td align="center">Tad_1, Tad_2</td>
</tr>
<tr class="odd">
<td align="right">2L</td>
<td align="center">100001</td>
<td align="center">105000</td>
<td align="center">*</td>
<td align="center">2L:21</td>
<td align="center">8</td>
<td align="center">active</td>
<td align="center">2L:21</td>
<td align="center">Tad_2, Tad_3</td>
</tr>
</tbody>
</table></details><p><strong>Step 3:</strong> This defines a GRanges object. In the
folowing examples, the same information is needed in a GInteraction
object class.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">domains_Border_Bin.gni</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu">InteractionSet</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/InteractionSet/man/GInteractions-class.html" class="external-link">GInteractions</a></span><span class="op">(</span></span>
<span>        <span class="va">domains_Border_Bin.gnr</span>,<span class="va">domains_Border_Bin.gnr</span><span class="op">)</span></span></code></pre></div>
<details><summary>
View
</summary><table class="table">
<thead>
<tr>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="8">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
First
</div>
</th>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="8">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
Second
</div>
</th>
</tr>
<tr>
<th style="text-align:right;">
seq
</th>
<th style="text-align:center;">
start
</th>
<th style="text-align:center;">
end
</th>
<th style="text-align:right;">
name
</th>
<th style="text-align:center;">
score
</th>
<th style="text-align:center;">
class
</th>
<th style="text-align:center;">
bin
</th>
<th style="text-align:center;">
subname
</th>
<th style="text-align:center;">
seq
</th>
<th style="text-align:center;">
start
</th>
<th style="text-align:center;">
end
</th>
<th style="text-align:center;">
name
</th>
<th style="text-align:center;">
score
</th>
<th style="text-align:center;">
class
</th>
<th style="text-align:center;">
bin
</th>
<th style="text-align:center;">
subname
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
70001
</td>
<td style="text-align:center;">
75000
</td>
<td style="text-align:right;">
2L:15
</td>
<td style="text-align:center;">
3
</td>
<td style="text-align:center;">
active
</td>
<td style="text-align:center;">
2L:15
</td>
<td style="text-align:center;">
Tad_1
</td>
<td style="text-align:center;">
2L
</td>
<td style="text-align:center;">
70001
</td>
<td style="text-align:center;">
75000
</td>
<td style="text-align:center;">
2L:15
</td>
<td style="text-align:center;">
active
</td>
<td style="text-align:center;">
3
</td>
<td style="text-align:center;">
2L:15
</td>
<td style="text-align:center;">
Tad_1
</td>
</tr>
<tr>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
90001
</td>
<td style="text-align:center;">
95000
</td>
<td style="text-align:right;">
2L:19
</td>
<td style="text-align:center;">
3, 8
</td>
<td style="text-align:center;">
active
</td>
<td style="text-align:center;">
2L:19
</td>
<td style="text-align:center;">
Tad_1, Tad_2
</td>
<td style="text-align:center;">
2L
</td>
<td style="text-align:center;">
90001
</td>
<td style="text-align:center;">
95000
</td>
<td style="text-align:center;">
2L:19
</td>
<td style="text-align:center;">
active
</td>
<td style="text-align:center;">
3, 8
</td>
<td style="text-align:center;">
2L:19
</td>
<td style="text-align:center;">
Tad_1, Tad_2
</td>
</tr>
<tr>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
100001
</td>
<td style="text-align:center;">
105000
</td>
<td style="text-align:right;">
2L:21
</td>
<td style="text-align:center;">
8
</td>
<td style="text-align:center;">
active
</td>
<td style="text-align:center;">
2L:21
</td>
<td style="text-align:center;">
Tad_2, Tad_3
</td>
<td style="text-align:center;">
2L
</td>
<td style="text-align:center;">
100001
</td>
<td style="text-align:center;">
105000
</td>
<td style="text-align:center;">
2L:21
</td>
<td style="text-align:center;">
active
</td>
<td style="text-align:center;">
8
</td>
<td style="text-align:center;">
2L:21
</td>
<td style="text-align:center;">
Tad_2, Tad_3
</td>
</tr>
</tbody>
</table></details>
</div>
<div class="section level4">
<h4 id="ponctual-interactions-defined-with-granges">Ponctual interactions defined with GRanges<a class="anchor" aria-label="anchor" href="#ponctual-interactions-defined-with-granges"></a>
</h4>
<p>Here the start and the end of each ranges are in a same bin.<br><img src="images/Ponctuals_Interactions_GRanges.png"><!-- --></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">border_PFmatrix.lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractSubmatrix.html">ExtractSubmatrix</a></span><span class="op">(</span></span>
<span>    genomicFeature         <span class="op">=</span> <span class="va">domains_Border_Bin.gnr</span>,</span>
<span>    hicLst        <span class="op">=</span> <span class="va">HiC_Ctrl.cmx_lst</span>,</span>
<span>    referencePoint <span class="op">=</span> <span class="st">"pf"</span>,</span>
<span>    matriceDim     <span class="op">=</span> <span class="fl">101</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="ponctual-interactions-defined-with-ginteractions">Ponctual interactions defined with GInteractions<a class="anchor" aria-label="anchor" href="#ponctual-interactions-defined-with-ginteractions"></a>
</h4>
<p>Here the first (blue on scheme) and the second (red on scheme)
elements are the same. <img src="images/Ponctuals_Interactions_GInteractions.png"><!-- --></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">border_PFmatrix.lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractSubmatrix.html">ExtractSubmatrix</a></span><span class="op">(</span></span>
<span>    genomicFeature         <span class="op">=</span> <span class="va">domains_Border_Bin.gni</span>,</span>
<span>    hicLst        <span class="op">=</span> <span class="va">HiC_Ctrl.cmx_lst</span>,</span>
<span>    referencePoint <span class="op">=</span> <span class="st">"pf"</span>,</span>
<span>    matriceDim     <span class="op">=</span> <span class="fl">101</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="tips-4">Tips<a class="anchor" aria-label="anchor" href="#tips-4"></a>
</h3>
<ol style="list-style-type: decimal">
<li>If <code>hicResolution</code> is NULL, the function will
atuomatically use the resolution of the <code>hicLst</code>
attributes.</li>
<li>referencePoint is automatically set as “pf” if every anchors and
baits are on the same bin (see examples).</li>
</ol>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="filtrations">Filtrations<a class="anchor" aria-label="anchor" href="#filtrations"></a>
</h2>
<p>The modularity of the workflow allows the user to filter
interactions, pairs or extracted submatrices at any step of the
analysis. <code>FilterInteractions</code> function takes as input either
a GInteraction object or a list of submatrices, and a list of targets of
choice and a selectionFunction defining how targets are filtered.</p>
<div class="section level3">
<h3 id="target-list-definition">Target list definition:<a class="anchor" aria-label="anchor" href="#target-list-definition"></a>
</h3>
<p>Target list must be defined by a named list corresponding to the same
names of each element and correspond to the column of the GInteraction
(or the attributes “interactions” of the matrices to be filtered). Then
each element must be a character list to match this column or a function
that will test each row in the column and return a bolean.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">structureTarget.lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    first_colname_of_GInteraction  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"value"</span><span class="op">)</span>,</span>
<span>    second_colname_of_GInteraction <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">eachElement</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="va">min_th</span><span class="op">&lt;</span><span class="va">value</span> <span class="op">&amp;&amp;</span> <span class="va">value</span><span class="op">&lt;</span><span class="va">max_th</span><span class="op">}</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>Interactions, pairs or extracted submatrices are filtered by metadata
elements from GRanges objects used in <code>SearchPairs</code>. Those
metadata are stored in the attributes of the list of submatrices that
are accessible as follow:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">interactions_RFmatrix_ctrl.lst</span><span class="op">)</span><span class="op">$</span><span class="va">interactions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">interactions_RFmatrix_ctrl.lst</span><span class="op">)</span><span class="op">$</span><span class="va">interactions</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<details><summary>
View
</summary><table class="table">
<thead>
<tr>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="6">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
Ranges
</div>
</th>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="15">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
Metadata
</div>
</th>
</tr>
<tr>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="3">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
First
</div>
</th>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="3">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
Second
</div>
</th>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="5">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
Interaction
</div>
</th>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="5">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
Anchor
</div>
</th>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="5">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
Bait
</div>
</th>
</tr>
<tr>
<th style="text-align:right;">
seq
</th>
<th style="text-align:center;">
start
</th>
<th style="text-align:center;">
end
</th>
<th style="text-align:right;">
seq
</th>
<th style="text-align:center;">
start
</th>
<th style="text-align:center;">
end
</th>
<th style="text-align:center;">
name
</th>
<th style="text-align:center;">
constraint
</th>
<th style="text-align:center;">
distance
</th>
<th style="text-align:center;">
orientation
</th>
<th style="text-align:center;">
submatrix.name
</th>
<th style="text-align:center;">
name
</th>
<th style="text-align:center;">
bin
</th>
<th style="text-align:center;">
Beaf.name
</th>
<th style="text-align:center;">
Beaf.score
</th>
<th style="text-align:center;">
Beaf.bln
</th>
<th style="text-align:center;">
name
</th>
<th style="text-align:center;">
bin
</th>
<th style="text-align:center;">
Tss.name
</th>
<th style="text-align:center;">
Tss.class
</th>
<th style="text-align:center;">
Tss.bln
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
120001
</td>
<td style="text-align:center;">
125000
</td>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
105001
</td>
<td style="text-align:center;">
110000
</td>
<td style="text-align:center;">
2L:25_2L:22
</td>
<td style="text-align:center;">
Tad_3
</td>
<td style="text-align:center;">
15000
</td>
<td style="text-align:center;">
FALSE
</td>
<td style="text-align:center;">
2L:22_2L:25
</td>
<td style="text-align:center;">
2L:25:Tad_3
</td>
<td style="text-align:center;">
2L:25
</td>
<td style="text-align:center;">
Beaf32_15
</td>
<td style="text-align:center;">
484
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
2L:22:Tad_3
</td>
<td style="text-align:center;">
2L:22
</td>
<td style="text-align:center;">
FBgn0026….
</td>
<td style="text-align:center;">
active
</td>
<td style="text-align:center;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
470001
</td>
<td style="text-align:center;">
475000
</td>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
420001
</td>
<td style="text-align:center;">
425000
</td>
<td style="text-align:center;">
2L:95_2L:85
</td>
<td style="text-align:center;">
Tad_17
</td>
<td style="text-align:center;">
50000
</td>
<td style="text-align:center;">
FALSE
</td>
<td style="text-align:center;">
2L:85_2L:95
</td>
<td style="text-align:center;">
2L:95:Tad_17
</td>
<td style="text-align:center;">
2L:95
</td>
<td style="text-align:center;">
Beaf32_62
</td>
<td style="text-align:center;">
37
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
2L:85:Tad_17
</td>
<td style="text-align:center;">
2L:85
</td>
<td style="text-align:center;">
FBgn0031253
</td>
<td style="text-align:center;">
active
</td>
<td style="text-align:center;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
470001
</td>
<td style="text-align:center;">
475000
</td>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
450001
</td>
<td style="text-align:center;">
455000
</td>
<td style="text-align:center;">
2L:95_2L:91
</td>
<td style="text-align:center;">
Tad_17
</td>
<td style="text-align:center;">
20000
</td>
<td style="text-align:center;">
FALSE
</td>
<td style="text-align:center;">
2L:91_2L:95
</td>
<td style="text-align:center;">
2L:95:Tad_17
</td>
<td style="text-align:center;">
2L:95
</td>
<td style="text-align:center;">
Beaf32_62
</td>
<td style="text-align:center;">
37
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
2L:91:Tad_17
</td>
<td style="text-align:center;">
2L:91
</td>
<td style="text-align:center;">
FBgn0015924
</td>
<td style="text-align:center;">
active
</td>
<td style="text-align:center;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
2490001
</td>
<td style="text-align:center;">
2495000
</td>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
2530001
</td>
<td style="text-align:center;">
2535000
</td>
<td style="text-align:center;">
2L:499_2L:507
</td>
<td style="text-align:center;">
Tad_64
</td>
<td style="text-align:center;">
40000
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
2L:499_2L:507
</td>
<td style="text-align:center;">
2L:499:Tad_64
</td>
<td style="text-align:center;">
2L:499
</td>
<td style="text-align:center;">
Beaf32_204
</td>
<td style="text-align:center;">
231
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
2L:507:Tad_64
</td>
<td style="text-align:center;">
2L:507
</td>
<td style="text-align:center;">
FBgn0264943
</td>
<td style="text-align:center;">
inactive
</td>
<td style="text-align:center;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
2675001
</td>
<td style="text-align:center;">
2680000
</td>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
2650001
</td>
<td style="text-align:center;">
2655000
</td>
<td style="text-align:center;">
2L:536_2L:531
</td>
<td style="text-align:center;">
Tad_67
</td>
<td style="text-align:center;">
25000
</td>
<td style="text-align:center;">
FALSE
</td>
<td style="text-align:center;">
2L:531_2L:536
</td>
<td style="text-align:center;">
2L:536:Tad_67
</td>
<td style="text-align:center;">
2L:536
</td>
<td style="text-align:center;">
Beaf32_210
</td>
<td style="text-align:center;">
124
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
2L:531:Tad_67
</td>
<td style="text-align:center;">
2L:531
</td>
<td style="text-align:center;">
FBgn0031442
</td>
<td style="text-align:center;">
inactive
</td>
<td style="text-align:center;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
2800001
</td>
<td style="text-align:center;">
2805000
</td>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
2770001
</td>
<td style="text-align:center;">
2775000
</td>
<td style="text-align:center;">
2L:561_2L:555
</td>
<td style="text-align:center;">
Tad_70
</td>
<td style="text-align:center;">
30000
</td>
<td style="text-align:center;">
FALSE
</td>
<td style="text-align:center;">
2L:555_2L:561
</td>
<td style="text-align:center;">
2L:561:Tad_70
</td>
<td style="text-align:center;">
2L:561
</td>
<td style="text-align:center;">
Beaf32_227
</td>
<td style="text-align:center;">
185
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
2L:555:Tad_70
</td>
<td style="text-align:center;">
2L:555
</td>
<td style="text-align:center;">
FBgn0019….
</td>
<td style="text-align:center;">
active
</td>
<td style="text-align:center;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
2805001
</td>
<td style="text-align:center;">
2810000
</td>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
2770001
</td>
<td style="text-align:center;">
2775000
</td>
<td style="text-align:center;">
2L:562_2L:555
</td>
<td style="text-align:center;">
Tad_70
</td>
<td style="text-align:center;">
35000
</td>
<td style="text-align:center;">
FALSE
</td>
<td style="text-align:center;">
2L:555_2L:562
</td>
<td style="text-align:center;">
2L:562:Tad_70
</td>
<td style="text-align:center;">
2L:562
</td>
<td style="text-align:center;">
Beaf32_227
</td>
<td style="text-align:center;">
185
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
2L:555:Tad_70
</td>
<td style="text-align:center;">
2L:555
</td>
<td style="text-align:center;">
FBgn0019….
</td>
<td style="text-align:center;">
active
</td>
<td style="text-align:center;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
2975001
</td>
<td style="text-align:center;">
2980000
</td>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
2990001
</td>
<td style="text-align:center;">
2995000
</td>
<td style="text-align:center;">
2L:596_2L:599
</td>
<td style="text-align:center;">
Tad_76
</td>
<td style="text-align:center;">
15000
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
2L:596_2L:599
</td>
<td style="text-align:center;">
2L:596:Tad_76
</td>
<td style="text-align:center;">
2L:596
</td>
<td style="text-align:center;">
Beaf32_244
</td>
<td style="text-align:center;">
98
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
2L:599:Tad_76
</td>
<td style="text-align:center;">
2L:599
</td>
<td style="text-align:center;">
FBgn0025681
</td>
<td style="text-align:center;">
active
</td>
<td style="text-align:center;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
3345001
</td>
<td style="text-align:center;">
3350000
</td>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
3330001
</td>
<td style="text-align:center;">
3335000
</td>
<td style="text-align:center;">
2L:670_2L:667
</td>
<td style="text-align:center;">
Tad_86
</td>
<td style="text-align:center;">
15000
</td>
<td style="text-align:center;">
FALSE
</td>
<td style="text-align:center;">
2L:667_2L:670
</td>
<td style="text-align:center;">
2L:670:Tad_86
</td>
<td style="text-align:center;">
2L:670
</td>
<td style="text-align:center;">
Beaf32_2….
</td>
<td style="text-align:center;">
179
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
2L:667:Tad_86
</td>
<td style="text-align:center;">
2L:667
</td>
<td style="text-align:center;">
FBgn0031523
</td>
<td style="text-align:center;">
inactive
</td>
<td style="text-align:center;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
3350001
</td>
<td style="text-align:center;">
3355000
</td>
<td style="text-align:right;">
2L
</td>
<td style="text-align:center;">
3330001
</td>
<td style="text-align:center;">
3335000
</td>
<td style="text-align:center;">
2L:671_2L:667
</td>
<td style="text-align:center;">
Tad_86
</td>
<td style="text-align:center;">
20000
</td>
<td style="text-align:center;">
FALSE
</td>
<td style="text-align:center;">
2L:667_2L:671
</td>
<td style="text-align:center;">
2L:671:Tad_86
</td>
<td style="text-align:center;">
2L:671
</td>
<td style="text-align:center;">
Beaf32_281
</td>
<td style="text-align:center;">
100
</td>
<td style="text-align:center;">
TRUE
</td>
<td style="text-align:center;">
2L:667:Tad_86
</td>
<td style="text-align:center;">
2L:667
</td>
<td style="text-align:center;">
FBgn0031523
</td>
<td style="text-align:center;">
inactive
</td>
<td style="text-align:center;">
TRUE
</td>
</tr>
</tbody>
</table></details><div class="section level4">
<h4 id="example-of-target-list">Example of target list:<a class="anchor" aria-label="anchor" href="#example-of-target-list"></a>
</h4>
<p>In this example, Pairs will be filtered on anchor.Beaf.name,
bait.Tss.name, name (which correponds to the submatrix IDs) and
distance. The aim of the example is to filter Pairs or submatrices that
have:</p>
<ol style="list-style-type: decimal">
<li>“Beaf32_62” and “Beaf32_204” in
<strong>anchor.Beaf.name</strong><br>
</li>
<li>“FBgn0015924” and “FBgn0264943” in
<strong>bait.Tss.name</strong><br>
</li>
<li>
<strong>distance</strong> exactly equal to 20000 or 40000 And to
exclude Pairs or submatrices that have:</li>
<li>“2L:25_2L:22” in <strong>name</strong>
</li>
</ol>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    anchor.Beaf.name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Beaf32_62"</span>,<span class="st">"Beaf32_204"</span><span class="op">)</span>,</span>
<span>    bait.Tss.name    <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FBgn0015924"</span>,<span class="st">"FBgn0264943"</span><span class="op">)</span>,</span>
<span>    name             <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"2L:25_2L:22"</span><span class="op">)</span>,</span>
<span>    distance         <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">columnElement</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fl">20000</span><span class="op">==</span><span class="va">columnElement</span> <span class="op">||</span> <span class="va">columnElement</span> <span class="op">==</span> <span class="fl">40000</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="selection-function-definition">Selection Function definition:<a class="anchor" aria-label="anchor" href="#selection-function-definition"></a>
</h3>
<p>The selectionFunction defines which operations (union(), intersect(),
setdiff()…) are used to filter the set of Pairs with target elements.
For more examples, see <em>Selection function tips and examples</em>
section.</p>
<div class="section level4">
<h4 id="example-of-selectionfunction-according-to-the-example-target">Example of selectionFunction according to the example target<a class="anchor" aria-label="anchor" href="#example-of-selectionfunction-according-to-the-example-target"></a>
</h4>
<p>Following the example case defined in targets</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selectionFun</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">intersect</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">anchor.Beaf.name</span>, <span class="va">bait.Tss.name</span> ,<span class="va">distance</span><span class="op">)</span> <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">name</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="filtration-with-selection">Filtration with selection<a class="anchor" aria-label="anchor" href="#filtration-with-selection"></a>
</h3>
<div class="section level4">
<h4 id="example-of-ginteraction-object-filtration">Example of GInteraction object filtration<a class="anchor" aria-label="anchor" href="#example-of-ginteraction-object-filtration"></a>
</h4>
<p>With a GInteraction object as input, <code>FilterInteractions</code>
will return the indices of filtered elements.</p>
<p>With the targets and selectionFun defined above:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/FilterInteractions.html">FilterInteractions</a></span><span class="op">(</span></span>
<span>    genomicInteractions <span class="op">=</span> </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">interactions_RFmatrix_ctrl.lst</span><span class="op">)</span><span class="op">$</span><span class="va">interactions</span>,</span>
<span>    targets        <span class="op">=</span> <span class="va">targets</span>,</span>
<span>    selectionFun     <span class="op">=</span> <span class="va">selectionFun</span></span>
<span>    <span class="op">)</span></span>
<span><span class="co">#&gt; [1] 3 4</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="example-of-matrices-list-filtration">Example of Matrices list filtration<a class="anchor" aria-label="anchor" href="#example-of-matrices-list-filtration"></a>
</h4>
<p>With a matrices list as input, <code>FilterInteractions</code> will
return the filtered matrices list, with updated attributes.</p>
<p>With the targets and selectionFun defined above:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filtred_interactions_RFmatrix_ctrl.lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterInteractions.html">FilterInteractions</a></span><span class="op">(</span></span>
<span>    matrices  <span class="op">=</span> <span class="va">interactions_RFmatrix_ctrl.lst</span>,</span>
<span>    targets    <span class="op">=</span> <span class="va">targets</span>,</span>
<span>    selectionFun <span class="op">=</span> <span class="va">selectionFun</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="specific-case-1-only-one-target-and-therefore-no-selection-needed">Specific case 1: Only one target (and therefore no selection
needed)<a class="anchor" aria-label="anchor" href="#specific-case-1-only-one-target-and-therefore-no-selection-needed"></a>
</h3>
<p>For example, to filter the top 100 first elements, select the top 100
first names</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">first100_targets</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    submatrix.name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">interactions_RFmatrix_ctrl.lst</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="ginteraction-filtration">GInteraction filtration<a class="anchor" aria-label="anchor" href="#ginteraction-filtration"></a>
</h4>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/FilterInteractions.html">FilterInteractions</a></span><span class="op">(</span></span>
<span>    genomicInteractions <span class="op">=</span> </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">interactions_RFmatrix_ctrl.lst</span><span class="op">)</span><span class="op">$</span><span class="va">interactions</span>,</span>
<span>    targets        <span class="op">=</span> <span class="va">first100_targets</span>,</span>
<span>    selectionFun     <span class="op">=</span> <span class="cn">NULL</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; submatrix.name1 submatrix.name2 submatrix.name3 submatrix.name4 submatrix.name5 </span></span>
<span><span class="co">#&gt;               1               2               3               4               5 </span></span>
<span><span class="co">#&gt; submatrix.name6 </span></span>
<span><span class="co">#&gt;               6</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="matrices-list-filtration">Matrices list filtration<a class="anchor" aria-label="anchor" href="#matrices-list-filtration"></a>
</h4>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">first100_interactions_RFmatrix_ctrl.lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterInteractions.html">FilterInteractions</a></span><span class="op">(</span></span>
<span>    matrices  <span class="op">=</span> <span class="va">interactions_RFmatrix_ctrl.lst</span>,</span>
<span>    targets    <span class="op">=</span> <span class="va">first100_targets</span>,</span>
<span>    selectionFun <span class="op">=</span> <span class="cn">NULL</span></span>
<span>    <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">first100_interactions_RFmatrix_ctrl.lst</span><span class="op">)</span><span class="op">$</span><span class="va">interactions</span></span>
<span><span class="co">#&gt; GInteractions object with 100 interactions and 15 metadata columns:</span></span>
<span><span class="co">#&gt;                   seqnames1           ranges1     seqnames2           ranges2 |</span></span>
<span><span class="co">#&gt;                       &lt;Rle&gt;         &lt;IRanges&gt;         &lt;Rle&gt;         &lt;IRanges&gt; |</span></span>
<span><span class="co">#&gt;       2L:25_2L:22        2L     120001-125000 ---        2L     105001-110000 |</span></span>
<span><span class="co">#&gt;       2L:95_2L:85        2L     470001-475000 ---        2L     420001-425000 |</span></span>
<span><span class="co">#&gt;       2L:95_2L:91        2L     470001-475000 ---        2L     450001-455000 |</span></span>
<span><span class="co">#&gt;     2L:499_2L:507        2L   2490001-2495000 ---        2L   2530001-2535000 |</span></span>
<span><span class="co">#&gt;     2L:536_2L:531        2L   2675001-2680000 ---        2L   2650001-2655000 |</span></span>
<span><span class="co">#&gt;               ...       ...               ... ...       ...               ... .</span></span>
<span><span class="co">#&gt;   2L:4313_2L:4299        2L 21560001-21565000 ---        2L 21490001-21495000 |</span></span>
<span><span class="co">#&gt;   2L:4315_2L:4299        2L 21570001-21575000 ---        2L 21490001-21495000 |</span></span>
<span><span class="co">#&gt;   2L:4312_2L:4305        2L 21555001-21560000 ---        2L 21520001-21525000 |</span></span>
<span><span class="co">#&gt;   2L:4313_2L:4305        2L 21560001-21565000 ---        2L 21520001-21525000 |</span></span>
<span><span class="co">#&gt;   2L:4315_2L:4305        2L 21570001-21575000 ---        2L 21520001-21525000 |</span></span>
<span><span class="co">#&gt;                              name  constraint  distance orientation</span></span>
<span><span class="co">#&gt;                       &lt;character&gt; &lt;character&gt; &lt;integer&gt;   &lt;logical&gt;</span></span>
<span><span class="co">#&gt;       2L:25_2L:22     2L:25_2L:22       Tad_3     15000       FALSE</span></span>
<span><span class="co">#&gt;       2L:95_2L:85     2L:95_2L:85      Tad_17     50000       FALSE</span></span>
<span><span class="co">#&gt;       2L:95_2L:91     2L:95_2L:91      Tad_17     20000       FALSE</span></span>
<span><span class="co">#&gt;     2L:499_2L:507   2L:499_2L:507      Tad_64     40000        TRUE</span></span>
<span><span class="co">#&gt;     2L:536_2L:531   2L:536_2L:531      Tad_67     25000       FALSE</span></span>
<span><span class="co">#&gt;               ...             ...         ...       ...         ...</span></span>
<span><span class="co">#&gt;   2L:4313_2L:4299 2L:4313_2L:4299     Tad_486     70000       FALSE</span></span>
<span><span class="co">#&gt;   2L:4315_2L:4299 2L:4315_2L:4299     Tad_486     80000       FALSE</span></span>
<span><span class="co">#&gt;   2L:4312_2L:4305 2L:4312_2L:4305     Tad_486     35000       FALSE</span></span>
<span><span class="co">#&gt;   2L:4313_2L:4305 2L:4313_2L:4305     Tad_486     40000       FALSE</span></span>
<span><span class="co">#&gt;   2L:4315_2L:4305 2L:4315_2L:4305     Tad_486     50000       FALSE</span></span>
<span><span class="co">#&gt;                    submatrix.name  anchor.bin     anchor.name    bait.bin</span></span>
<span><span class="co">#&gt;                       &lt;character&gt; &lt;character&gt;     &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;       2L:25_2L:22     2L:22_2L:25       2L:25     2L:25:Tad_3       2L:22</span></span>
<span><span class="co">#&gt;       2L:95_2L:85     2L:85_2L:95       2L:95    2L:95:Tad_17       2L:85</span></span>
<span><span class="co">#&gt;       2L:95_2L:91     2L:91_2L:95       2L:95    2L:95:Tad_17       2L:91</span></span>
<span><span class="co">#&gt;     2L:499_2L:507   2L:499_2L:507      2L:499   2L:499:Tad_64      2L:507</span></span>
<span><span class="co">#&gt;     2L:536_2L:531   2L:531_2L:536      2L:536   2L:536:Tad_67      2L:531</span></span>
<span><span class="co">#&gt;               ...             ...         ...             ...         ...</span></span>
<span><span class="co">#&gt;   2L:4313_2L:4299 2L:4299_2L:4313     2L:4313 2L:4313:Tad_486     2L:4299</span></span>
<span><span class="co">#&gt;   2L:4315_2L:4299 2L:4299_2L:4315     2L:4315 2L:4315:Tad_486     2L:4299</span></span>
<span><span class="co">#&gt;   2L:4312_2L:4305 2L:4305_2L:4312     2L:4312 2L:4312:Tad_486     2L:4305</span></span>
<span><span class="co">#&gt;   2L:4313_2L:4305 2L:4305_2L:4313     2L:4313 2L:4313:Tad_486     2L:4305</span></span>
<span><span class="co">#&gt;   2L:4315_2L:4305 2L:4305_2L:4315     2L:4315 2L:4315:Tad_486     2L:4305</span></span>
<span><span class="co">#&gt;                         bait.name anchor.Beaf.score anchor.Beaf.name</span></span>
<span><span class="co">#&gt;                       &lt;character&gt;         &lt;numeric&gt;           &lt;list&gt;</span></span>
<span><span class="co">#&gt;       2L:25_2L:22     2L:22:Tad_3               484        Beaf32_15</span></span>
<span><span class="co">#&gt;       2L:95_2L:85    2L:85:Tad_17                37        Beaf32_62</span></span>
<span><span class="co">#&gt;       2L:95_2L:91    2L:91:Tad_17                37        Beaf32_62</span></span>
<span><span class="co">#&gt;     2L:499_2L:507   2L:507:Tad_64               231       Beaf32_204</span></span>
<span><span class="co">#&gt;     2L:536_2L:531   2L:531:Tad_67               124       Beaf32_210</span></span>
<span><span class="co">#&gt;               ...             ...               ...              ...</span></span>
<span><span class="co">#&gt;   2L:4313_2L:4299 2L:4299:Tad_486               748      Beaf32_1348</span></span>
<span><span class="co">#&gt;   2L:4315_2L:4299 2L:4299:Tad_486               529      Beaf32_1349</span></span>
<span><span class="co">#&gt;   2L:4312_2L:4305 2L:4305:Tad_486               748      Beaf32_1348</span></span>
<span><span class="co">#&gt;   2L:4313_2L:4305 2L:4305:Tad_486               748      Beaf32_1348</span></span>
<span><span class="co">#&gt;   2L:4315_2L:4305 2L:4305:Tad_486               529      Beaf32_1349</span></span>
<span><span class="co">#&gt;                   anchor.Beaf.bln bait.Tss.class           bait.Tss.name</span></span>
<span><span class="co">#&gt;                         &lt;logical&gt;         &lt;list&gt;                  &lt;list&gt;</span></span>
<span><span class="co">#&gt;       2L:25_2L:22            TRUE         active FBgn0026787,FBgn0005278</span></span>
<span><span class="co">#&gt;       2L:95_2L:85            TRUE         active             FBgn0031253</span></span>
<span><span class="co">#&gt;       2L:95_2L:91            TRUE         active             FBgn0015924</span></span>
<span><span class="co">#&gt;     2L:499_2L:507            TRUE       inactive             FBgn0264943</span></span>
<span><span class="co">#&gt;     2L:536_2L:531            TRUE       inactive             FBgn0031442</span></span>
<span><span class="co">#&gt;               ...             ...            ...                     ...</span></span>
<span><span class="co">#&gt;   2L:4313_2L:4299            TRUE       inactive             FBgn0053837</span></span>
<span><span class="co">#&gt;   2L:4315_2L:4299            TRUE       inactive             FBgn0053837</span></span>
<span><span class="co">#&gt;   2L:4312_2L:4305            TRUE       inactive             FBgn0053873</span></span>
<span><span class="co">#&gt;   2L:4313_2L:4305            TRUE       inactive             FBgn0053873</span></span>
<span><span class="co">#&gt;   2L:4315_2L:4305            TRUE       inactive             FBgn0053873</span></span>
<span><span class="co">#&gt;                   bait.Tss.bln</span></span>
<span><span class="co">#&gt;                      &lt;logical&gt;</span></span>
<span><span class="co">#&gt;       2L:25_2L:22         TRUE</span></span>
<span><span class="co">#&gt;       2L:95_2L:85         TRUE</span></span>
<span><span class="co">#&gt;       2L:95_2L:91         TRUE</span></span>
<span><span class="co">#&gt;     2L:499_2L:507         TRUE</span></span>
<span><span class="co">#&gt;     2L:536_2L:531         TRUE</span></span>
<span><span class="co">#&gt;               ...          ...</span></span>
<span><span class="co">#&gt;   2L:4313_2L:4299         TRUE</span></span>
<span><span class="co">#&gt;   2L:4315_2L:4299         TRUE</span></span>
<span><span class="co">#&gt;   2L:4312_2L:4305         TRUE</span></span>
<span><span class="co">#&gt;   2L:4313_2L:4305         TRUE</span></span>
<span><span class="co">#&gt;   2L:4315_2L:4305         TRUE</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 335 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome</span></span></code></pre></div>
<p><strong>Warning!</strong> A selection of some matrices removes
attributes.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">interactions_RFmatrix_ctrl.lst</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">interactions</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="specific-case-2-sampling">Specific case 2: Sampling<a class="anchor" aria-label="anchor" href="#specific-case-2-sampling"></a>
</h3>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nSample.num</span> <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">targets</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>name<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">interactions_RFmatrix_ctrl.lst</span><span class="op">)</span><span class="op">$</span><span class="va">interactions</span><span class="op">$</span><span class="va">name</span>,<span class="va">nSample.num</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="ginteraction-sampling">GInteraction sampling<a class="anchor" aria-label="anchor" href="#ginteraction-sampling"></a>
</h4>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/FilterInteractions.html">FilterInteractions</a></span><span class="op">(</span></span>
<span>    genomicInteractions <span class="op">=</span> </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">interactions_RFmatrix_ctrl.lst</span><span class="op">)</span><span class="op">$</span><span class="va">interactions</span>,</span>
<span>    targets        <span class="op">=</span> <span class="va">targets</span>,</span>
<span>    selectionFun     <span class="op">=</span> <span class="cn">NULL</span></span>
<span>    <span class="op">)</span></span>
<span><span class="co">#&gt; name1 name2 name3 </span></span>
<span><span class="co">#&gt;    14    50   118</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="matrices-list-sampling">Matrices list sampling<a class="anchor" aria-label="anchor" href="#matrices-list-sampling"></a>
</h4>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampled_interactions_RFmatrix_ctrl.lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterInteractions.html">FilterInteractions</a></span><span class="op">(</span></span>
<span>    matrices  <span class="op">=</span> <span class="va">interactions_RFmatrix_ctrl.lst</span>,</span>
<span>    targets    <span class="op">=</span> <span class="va">targets</span>,</span>
<span>    selectionFun <span class="op">=</span> <span class="cn">NULL</span></span>
<span>    <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">sampled_interactions_RFmatrix_ctrl.lst</span><span class="op">)</span><span class="op">$</span><span class="va">interactions</span></span>
<span><span class="co">#&gt; GInteractions object with 3 interactions and 15 metadata columns:</span></span>
<span><span class="co">#&gt;                   seqnames1           ranges1     seqnames2           ranges2 |</span></span>
<span><span class="co">#&gt;                       &lt;Rle&gt;         &lt;IRanges&gt;         &lt;Rle&gt;         &lt;IRanges&gt; |</span></span>
<span><span class="co">#&gt;   2L:1012_2L:1015        2L   5055001-5060000 ---        2L   5070001-5075000 |</span></span>
<span><span class="co">#&gt;   2L:3002_2L:2978        2L 15005001-15010000 ---        2L 14885001-14890000 |</span></span>
<span><span class="co">#&gt;   2R:1824_2R:1820        2R   9115001-9120000 ---        2R   9095001-9100000 |</span></span>
<span><span class="co">#&gt;                              name  constraint  distance orientation</span></span>
<span><span class="co">#&gt;                       &lt;character&gt; &lt;character&gt; &lt;integer&gt;   &lt;logical&gt;</span></span>
<span><span class="co">#&gt;   2L:1012_2L:1015 2L:1012_2L:1015     Tad_132     15000        TRUE</span></span>
<span><span class="co">#&gt;   2L:3002_2L:2978 2L:3002_2L:2978     Tad_356    120000       FALSE</span></span>
<span><span class="co">#&gt;   2R:1824_2R:1820 2R:1824_2R:1820     Tad_608     20000       FALSE</span></span>
<span><span class="co">#&gt;                    submatrix.name  anchor.bin     anchor.name    bait.bin</span></span>
<span><span class="co">#&gt;                       &lt;character&gt; &lt;character&gt;     &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt;   2L:1012_2L:1015 2L:1012_2L:1015     2L:1012 2L:1012:Tad_132     2L:1015</span></span>
<span><span class="co">#&gt;   2L:3002_2L:2978 2L:2978_2L:3002     2L:3002 2L:3002:Tad_356     2L:2978</span></span>
<span><span class="co">#&gt;   2R:1824_2R:1820 2R:1820_2R:1824     2R:1824 2R:1824:Tad_608     2R:1820</span></span>
<span><span class="co">#&gt;                         bait.name anchor.Beaf.score      anchor.Beaf.name</span></span>
<span><span class="co">#&gt;                       &lt;character&gt;         &lt;numeric&gt;                &lt;list&gt;</span></span>
<span><span class="co">#&gt;   2L:1012_2L:1015 2L:1015:Tad_132               660 Beaf32_374,Beaf32_375</span></span>
<span><span class="co">#&gt;   2L:3002_2L:2978 2L:2978:Tad_356                49           Beaf32_1001</span></span>
<span><span class="co">#&gt;   2R:1824_2R:1820 2R:1820:Tad_608               977           Beaf32_1734</span></span>
<span><span class="co">#&gt;                   anchor.Beaf.bln bait.Tss.class bait.Tss.name bait.Tss.bln</span></span>
<span><span class="co">#&gt;                         &lt;logical&gt;         &lt;list&gt;        &lt;list&gt;    &lt;logical&gt;</span></span>
<span><span class="co">#&gt;   2L:1012_2L:1015            TRUE         active   FBgn0261608         TRUE</span></span>
<span><span class="co">#&gt;   2L:3002_2L:2978            TRUE       inactive   FBgn0266840         TRUE</span></span>
<span><span class="co">#&gt;   2R:1824_2R:1820            TRUE         active   FBgn0082927         TRUE</span></span>
<span><span class="co">#&gt;   -------</span></span>
<span><span class="co">#&gt;   regions: 335 ranges and 0 metadata columns</span></span>
<span><span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="specific-case-3-filtration-without-selectionfunction">Specific case 3: Filtration without selectionFunction<a class="anchor" aria-label="anchor" href="#specific-case-3-filtration-without-selectionfunction"></a>
</h3>
<p>Without any selectionFunction, <code>FilterInteractions</code> will
return all indices corresponding to each target in the list. Then, the
indices of interest can be selected in a second step. For the examples
we take the folowing targets:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    anchor.Beaf.name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Beaf32_8"</span>,<span class="st">"Beaf32_15"</span><span class="op">)</span>,</span>
<span>    bait.Tss.name    <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FBgn0031214"</span>,<span class="st">"FBgn0005278"</span><span class="op">)</span>,</span>
<span>    name             <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"2L:74_2L:77"</span><span class="op">)</span>,</span>
<span>    distance         <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">columnElement</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fl">14000</span><span class="op">==</span><span class="va">columnElement</span> <span class="op">||</span> <span class="va">columnElement</span> <span class="op">==</span> <span class="fl">3000</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="ginteraction-filtration-1">GInteraction filtration<a class="anchor" aria-label="anchor" href="#ginteraction-filtration-1"></a>
</h4>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/FilterInteractions.html">FilterInteractions</a></span><span class="op">(</span></span>
<span>    genomicInteractions <span class="op">=</span> </span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">interactions_RFmatrix_ctrl.lst</span><span class="op">)</span><span class="op">$</span><span class="va">interactions</span>,</span>
<span>    targets        <span class="op">=</span> <span class="va">targets</span>,</span>
<span>    selectionFun     <span class="op">=</span> <span class="cn">NULL</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 4</span></span>
<span><span class="co">#&gt;  $ anchor.Beaf.name: int 1</span></span>
<span><span class="co">#&gt;  $ bait.Tss.name   : int 1</span></span>
<span><span class="co">#&gt;  $ name            : int(0) </span></span>
<span><span class="co">#&gt;  $ distance        : int(0)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="matrices-list-filtration-1">Matrices list filtration<a class="anchor" aria-label="anchor" href="#matrices-list-filtration-1"></a>
</h4>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/FilterInteractions.html">FilterInteractions</a></span><span class="op">(</span></span>
<span>    matrices      <span class="op">=</span> <span class="va">interactions_RFmatrix_ctrl.lst</span>,</span>
<span>    targets        <span class="op">=</span> <span class="va">targets</span>,</span>
<span>    selectionFun     <span class="op">=</span> <span class="cn">NULL</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 4</span></span>
<span><span class="co">#&gt;  $ anchor.Beaf.name: int 1</span></span>
<span><span class="co">#&gt;  $ bait.Tss.name   : int 1</span></span>
<span><span class="co">#&gt;  $ name            : int(0) </span></span>
<span><span class="co">#&gt;  $ distance        : int(0)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="tips-5">Tips<a class="anchor" aria-label="anchor" href="#tips-5"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Filter a GInteraction object allows to intersect the selected
index.<br>
</li>
<li>Filter a matrices list without selection is better than filter the
interaction attributes of the matrices list</li>
</ol>
<div class="section level4">
<h4 id="selection-function-tips-and-examples">Selection function tips and examples:<a class="anchor" aria-label="anchor" href="#selection-function-tips-and-examples"></a>
</h4>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>,<span class="st">"B"</span>,<span class="st">"D"</span>,<span class="st">"G"</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"E"</span>,<span class="st">"B"</span>,<span class="st">"C"</span>,<span class="st">"G"</span><span class="op">)</span></span>
<span><span class="va">c</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>,<span class="st">"F"</span>,<span class="st">"C"</span>,<span class="st">"G"</span><span class="op">)</span></span></code></pre></div>
<ol style="list-style-type: decimal">
<li>What is common to A, B and C</li>
</ol>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">intersect</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "G"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">c</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "G"</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>What is in A and/or B and/or C</li>
</ol>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">union</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "A" "B" "C" "D" "E" "F" "G"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">union</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">union</a></span><span class="op">(</span><span class="va">c</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "A" "B" "C" "D" "E" "F" "G"</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>What is only in A</li>
</ol>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="va">setdiff</span>,<span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "D"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">c</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "D"</span></span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li>What is common in A with B, and not in C</li>
</ol>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">c</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "B"</span></span></code></pre></div>
<ol start="5" style="list-style-type: decimal">
<li>What is common in A with B, plus all that is present in C</li>
</ol>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">union</a></span><span class="op">(</span><span class="va">c</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "A" "B" "C" "F" "G"</span></span></code></pre></div>
<ol start="6" style="list-style-type: decimal">
<li>What is common in C with all elements present in A and B</li>
</ol>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">union</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">c</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "A" "C" "G"</span></span></code></pre></div>
<ol start="7" style="list-style-type: decimal">
<li>Everything that is present in A and B but not in C</li>
</ol>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">union</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">c</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "B" "D" "E"</span></span></code></pre></div>
<ol start="8" style="list-style-type: decimal">
<li>What is present only once</li>
</ol>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">b</span>,<span class="va">c</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">d</span>,<span class="va">d</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "D" "E" "F"</span></span></code></pre></div>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="orientation">Orientation<a class="anchor" aria-label="anchor" href="#orientation"></a>
</h2>
<p><code>ExtractSubmatrix</code> returns submatrices orientated
according to 5’-&gt;3’ orientation of the chromosome. In the case of
heterotypic or asymetric pairs (anchor != bait), anchors and baits are
thus mixed on Y and X axis of the matrices.</p>
<p><img src="images/Orientation_extraction.png"><!-- --></p>
<p><code>OrientateMatrix</code> function allows to force all matrices to
be orientated in a way that anchors will be systematically on Y axis and
baits on X axis.</p>
<p><img src="images/Orientation.png"><!-- --></p>
<div class="section level3">
<h3 id="information-about-the-orientation">Information about the orientation<a class="anchor" aria-label="anchor" href="#information-about-the-orientation"></a>
</h3>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># mcols(attributes(</span></span>
<span>    <span class="co"># first100_interactions_RFmatrix_ctrl.lst)$interactions)$orientation</span></span></code></pre></div>
<ol style="list-style-type: decimal">
<li>The 13th matrice is well oriented, i.e. the <strong>anchor Beaf is
in Y axis</strong> and the <strong>bait TSS in X axis</strong><br>
</li>
<li>The 14th matrice is not well oriented, i.e. the <strong>bait TSS is
in Y axis</strong> and the <strong>anchor Beaf in X axis</strong>
</li>
</ol>
</div>
<div class="section level3">
<h3 id="orientation-on-matrices-list">Orientation on matrices list<a class="anchor" aria-label="anchor" href="#orientation-on-matrices-list"></a>
</h3>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oriented_first100_interactions_RFmatrix_ctrl.lst</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu"><a href="../reference/OrientateMatrix.html">OrientateMatrix</a></span><span class="op">(</span><span class="va">first100_interactions_RFmatrix_ctrl.lst</span><span class="op">)</span></span>
<span><span class="co">#&gt; 69 matrices are oriented</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="orientation-of-one-matrix-only-">Orientation of one matrix only.<a class="anchor" aria-label="anchor" href="#orientation-of-one-matrix-only-"></a>
</h3>
<p><strong>Warning</strong> This procedure force orientation even if not
needed.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">orientedMatrix.mtx</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu"><a href="../reference/OrientateMatrix.html">OrientateMatrix</a></span><span class="op">(</span><span class="va">first100_interactions_RFmatrix_ctrl.lst</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="prepare-matrices-list">Prepare matrices list<a class="anchor" aria-label="anchor" href="#prepare-matrices-list"></a>
</h2>
<p><code>PrepareMtxList</code> can be used to perform operations on the
matrices list. This function prepares the matrices list by performing a
per matrix operation by transforming values. For example values can be
quantilized inorder to rank local interactions and highlight on the
contacts with the highest values. This function can also be used to
correct orientations, just as <code>OrientateMatrix</code>. The reason
for giving access for the user to this originally hidden function is to
have a uniformly prepared matrices list for both quantifications and
visualisations.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oriented_quantiled_first100_interactions_RFmatrix_ctrl.lst</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu"><a href="../reference/PrepareMtxList.html">PrepareMtxList</a></span><span class="op">(</span></span>
<span>        <span class="va">first100_interactions_RFmatrix_ctrl.lst</span>,</span>
<span>        transFun <span class="op">=</span> <span class="st">'quantile'</span>,</span>
<span>        orientate <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; 69 matrices are oriented</span></span>
<span><span class="va">oriented_first100_interactions_RFmatrix_ctrl.lst</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu"><a href="../reference/PrepareMtxList.html">PrepareMtxList</a></span><span class="op">(</span></span>
<span>        <span class="va">first100_interactions_RFmatrix_ctrl.lst</span>,</span>
<span>        orientate <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; 69 matrices are oriented</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="quantifications">Quantifications<a class="anchor" aria-label="anchor" href="#quantifications"></a>
</h2>
<p><code>GetQuantif</code> function takes as input a list of submatrices
and returns a vector of contact frequencies in a given typeof regions
where these contacts are computed with a function.</p>
<div class="section level3">
<h3 id="basic-quantifications">Basic quantifications<a class="anchor" aria-label="anchor" href="#basic-quantifications"></a>
</h3>
<p>The <code>GetQuantif</code> function extracts per submatrix the
average values of the 3*3 central pixels by default (see
<code>GetQuantif</code>).</p>
<ol style="list-style-type: decimal">
<li>
<strong>area</strong>: The region where the contacts values are
extracted in each matrix.</li>
<li>
<strong>operation</strong> The function that is done on extracted
values for each matrix.</li>
</ol>
<p>Example: Average of the values in the centered 3x3 square.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">center.num</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetQuantif.html">GetQuantif</a></span><span class="op">(</span></span>
<span>    matrices  <span class="op">=</span> <span class="va">oriented_first100_interactions_RFmatrix_ctrl.lst</span>,</span>
<span>    areaFun      <span class="op">=</span> <span class="st">"center"</span>,</span>
<span>    operationFun <span class="op">=</span> <span class="st">"mean"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="custom-functions">Custom functions<a class="anchor" aria-label="anchor" href="#custom-functions"></a>
</h3>
<p>The <code>GetQuantif</code> function also takes custom
<strong>area</strong> and <strong>operation</strong> in parameter.</p>
<ol style="list-style-type: decimal">
<li>
<strong>area</strong>: function defining on which submatrix
coordinates the values are extracted in each matrices.<br>
</li>
<li>
<strong>operation</strong> function defining which operation is done
on extracted values for each matrices.</li>
</ol>
<p>Example: Interactions values on the matrice.mtx[33:35,67:69] area,
averaged after removing all zeros.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/GetQuantif.html">GetQuantif</a></span><span class="op">(</span></span>
<span>    matrices  <span class="op">=</span> <span class="va">oriented_first100_interactions_RFmatrix_ctrl.lst</span>,</span>
<span>    areaFun      <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">matrice.mtx</span><span class="op">)</span><span class="op">{</span><span class="va">matrice.mtx</span><span class="op">[</span><span class="fl">33</span><span class="op">:</span><span class="fl">35</span>,<span class="fl">67</span><span class="op">:</span><span class="fl">69</span><span class="op">]</span><span class="op">}</span>,</span>
<span>    operationFun <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">area.mtx</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="va">area.mtx</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">area.mtx</span><span class="op">==</span><span class="fl">0</span><span class="op">)</span><span class="op">]</span><span class="op">&lt;-</span><span class="cn">NA</span>;</span>
<span>        <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">area.mtx</span>,na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   2L:25_2L:22   2L:95_2L:85   2L:95_2L:91 2L:499_2L:507 2L:536_2L:531 </span></span>
<span><span class="co">#&gt;     1.2652268     1.0275631     1.8478690     0.8648371     1.1975185 </span></span>
<span><span class="co">#&gt; 2L:561_2L:555 </span></span>
<span><span class="co">#&gt;     1.6126441</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="particular-cases">Particular cases:<a class="anchor" aria-label="anchor" href="#particular-cases"></a>
</h3>
<div class="section level4">
<h4 id="values-naming">Values naming<a class="anchor" aria-label="anchor" href="#values-naming"></a>
</h4>
<p>By default, returned values are named with submatrix ID. If
<code>varName</code> is set with an element metadata column name from
GInteraction attributes, values are returned values are named according
to this element.</p>
<p>Example: Named quantifications with <code>anchor.Beaf.name</code></p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">namedCenter.num</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetQuantif.html">GetQuantif</a></span><span class="op">(</span></span>
<span>    matrices  <span class="op">=</span> <span class="va">oriented_first100_interactions_RFmatrix_ctrl.lst</span>,</span>
<span>    areaFun      <span class="op">=</span> <span class="st">"center"</span>,</span>
<span>    operationFun <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>    varName      <span class="op">=</span> <span class="st">"anchor.Beaf.name"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p>Note that changing submatrix ID for other names can create name
duplicates:</p>
<p>Example: The 46th matrix is correspond to two Beaf32 peaks, i.e. it
has two anchor.Beaf.name</p>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="center">name</th>
<th align="center">anchor.Beaf.name</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">45</td>
<td align="center">2L:2676_2L:2680</td>
<td align="center">Beaf32_950</td>
</tr>
<tr class="even">
<td align="left">46</td>
<td align="center">2L:2768_2L:2765</td>
<td align="center">Beaf32_981</td>
</tr>
<tr class="odd">
<td align="left">47</td>
<td align="center">2L:2971_2L:2977</td>
<td align="center">Beaf32_1000</td>
</tr>
<tr class="even">
<td align="left">48</td>
<td align="center">2L:3002_2L:2977</td>
<td align="center">Beaf32_1001</td>
</tr>
<tr class="odd">
<td align="left">49</td>
<td align="center">2L:2971_2L:2978</td>
<td align="center">Beaf32_1000</td>
</tr>
<tr class="even">
<td align="left">50</td>
<td align="center">2L:3002_2L:2978</td>
<td align="center">Beaf32_1001</td>
</tr>
</tbody>
</table>
<p>As a consequence, the value in center.num is duplicated in
namedCenter.num</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">center.num</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">45</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span></span>
<span><span class="co">#&gt; 2L:2676_2L:2680 2L:2768_2L:2765 2L:2971_2L:2977 2L:3002_2L:2977 2L:2971_2L:2978 </span></span>
<span><span class="co">#&gt;       0.7318490       0.8071858       0.5238616       0.7522744       0.5438497 </span></span>
<span><span class="co">#&gt; 2L:3002_2L:2978 </span></span>
<span><span class="co">#&gt;       0.7135216</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">namedCenter.num</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">45</span><span class="op">:</span><span class="fl">51</span><span class="op">]</span></span>
<span><span class="co">#&gt; Beaf32_378 Beaf32_408 Beaf32_408 Beaf32_408 Beaf32_437 Beaf32_518 Beaf32_521 </span></span>
<span><span class="co">#&gt;  1.1762447  1.0476893  0.8868763  0.7318490  0.8071858  0.5238616  0.7522744</span></span></code></pre></div>
<p>Duplicated value index are stored in attributes.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">center.num</span><span class="op">)</span><span class="op">$</span><span class="va">duplicated</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attributes.html" class="external-link">attributes</a></span><span class="op">(</span><span class="va">namedCenter.num</span><span class="op">)</span><span class="op">$</span><span class="va">duplicated</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="one-value-extraction">One value extraction<a class="anchor" aria-label="anchor" href="#one-value-extraction"></a>
</h4>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/GetQuantif.html">GetQuantif</a></span><span class="op">(</span></span>
<span>    matrices  <span class="op">=</span> <span class="va">oriented_first100_interactions_RFmatrix_ctrl.lst</span>,</span>
<span>    areaFun      <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">matrice.mtx</span><span class="op">)</span><span class="op">{</span><span class="va">matrice.mtx</span><span class="op">[</span><span class="fl">5</span>,<span class="fl">5</span><span class="op">]</span><span class="op">}</span>,</span>
<span>    operationFun <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">area.mtx</span><span class="op">)</span><span class="op">{</span><span class="va">area.mtx</span><span class="op">}</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   2L:25_2L:22   2L:95_2L:85   2L:95_2L:91 2L:499_2L:507 2L:536_2L:531 </span></span>
<span><span class="co">#&gt;     1.6036209     1.2231158     1.2129980     0.8516612     0.3858703 </span></span>
<span><span class="co">#&gt; 2L:561_2L:555 </span></span>
<span><span class="co">#&gt;     0.8043792</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="area-extraction">Area extraction<a class="anchor" aria-label="anchor" href="#area-extraction"></a>
</h4>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/GetQuantif.html">GetQuantif</a></span><span class="op">(</span></span>
<span>    matrices  <span class="op">=</span> <span class="va">oriented_first100_interactions_RFmatrix_ctrl.lst</span>,</span>
<span>    areaFun      <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">matrice.mtx</span><span class="op">)</span><span class="op">{</span><span class="va">matrice.mtx</span><span class="op">[</span><span class="fl">4</span><span class="op">:</span><span class="fl">6</span>,<span class="fl">4</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">}</span>,</span>
<span>    operationFun <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">area</span><span class="op">)</span><span class="op">{</span><span class="va">area</span><span class="op">}</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   2L:25_2L:22   2L:95_2L:85   2L:95_2L:91 2L:499_2L:507 2L:536_2L:531 </span></span>
<span><span class="co">#&gt;      1.599228            NA            NA      1.716875      1.603621 </span></span>
<span><span class="co">#&gt; 2L:561_2L:555 </span></span>
<span><span class="co">#&gt;            NA</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="tips-6">Tips<a class="anchor" aria-label="anchor" href="#tips-6"></a>
</h3>
<ol style="list-style-type: decimal">
<li>If <code>operationFun</code> is NULL then it will return values of
the selected region without NA.</li>
</ol>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/GetQuantif.html">GetQuantif</a></span><span class="op">(</span></span>
<span>    matrices  <span class="op">=</span> <span class="va">oriented_first100_interactions_RFmatrix_ctrl.lst</span>,</span>
<span>    areaFun      <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">matrice.mtx</span><span class="op">)</span><span class="op">{</span><span class="va">matrice.mtx</span><span class="op">[</span><span class="fl">4</span><span class="op">:</span><span class="fl">6</span>,<span class="fl">4</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">}</span>,</span>
<span>    operationFun <span class="op">=</span> <span class="cn">NULL</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   2L:25_2L:22   2L:95_2L:85   2L:95_2L:91 2L:499_2L:507 2L:536_2L:531 </span></span>
<span><span class="co">#&gt;      1.599228      1.716875      1.603621      1.834523      1.731368 </span></span>
<span><span class="co">#&gt; 2L:561_2L:555 </span></span>
<span><span class="co">#&gt;      1.628212</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="aggregations">Aggregations<a class="anchor" aria-label="anchor" href="#aggregations"></a>
</h2>
<p><code>Aggregation</code> function takes as input a list of
submatrices and returns an aggregated matrix using the aggregation
function defined by the user.</p>
<div class="section level3">
<h3 id="one-sample-aggregation">One sample aggregation<a class="anchor" aria-label="anchor" href="#one-sample-aggregation"></a>
</h3>
<div class="section level4">
<h4 id="basic-aggregation">Basic aggregation<a class="anchor" aria-label="anchor" href="#basic-aggregation"></a>
</h4>
<p><code>Aggregation</code> function has some default aggregation
functions like <code>sum</code>, <code>mean</code> or
<code>median</code> (see <code>Aggregation</code>)</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rm0 argument can be added to PrepareMtxList to assign NA to 0 values.</span></span>
<span><span class="va">oriented_first100_interactions_RFmatrix_ctrl.lst</span> <span class="op">=</span> </span>
<span>    <span class="fu"><a href="../reference/PrepareMtxList.html">PrepareMtxList</a></span><span class="op">(</span></span>
<span>        <span class="va">oriented_first100_interactions_RFmatrix_ctrl.lst</span>,</span>
<span>        rm0 <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">agg_sum.mtx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Aggregation.html">Aggregation</a></span><span class="op">(</span></span>
<span>    matrices <span class="op">=</span> <span class="va">oriented_first100_interactions_RFmatrix_ctrl.lst</span>, </span>
<span>    aggFun      <span class="op">=</span> <span class="st">"sum"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="custom-aggregation">Custom aggregation<a class="anchor" aria-label="anchor" href="#custom-aggregation"></a>
</h4>
<p>Defining a custom aggregation function: example below shows the mean
function after removing NA.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">agg_mean.mtx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Aggregation.html">Aggregation</a></span><span class="op">(</span></span>
<span>    matrices <span class="op">=</span> <span class="va">oriented_first100_interactions_RFmatrix_ctrl.lst</span>,</span>
<span>    aggFun      <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">x</span>,na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">}</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="two-samples-differential-aggregation">Two samples differential aggregation<a class="anchor" aria-label="anchor" href="#two-samples-differential-aggregation"></a>
</h3>
<p><code>Aggregation</code> function can take as input two list of
submatrices from two samples or conditions and returns a differential
aggregated matrix. Two ways to obtain differential aggregation are
applied, first is by assessing differences on each individual pairs of
submatrices then aggregate the differences; second is by aggregating
matrices and assess differences on the aggregated matrices (see examples
below).</p>
<div class="section level4">
<h4 id="preparation-of-matrices-list">Preparation of matrices list<a class="anchor" aria-label="anchor" href="#preparation-of-matrices-list"></a>
</h4>
<ol style="list-style-type: decimal">
<li>Preparation of Control matrices list condition<br><strong>Filtration</strong>
</li>
</ol>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">first100_targets</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    submatrix.name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">interactions_RFmatrix_ctrl.lst</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span>
<span>    <span class="op">)</span></span>
<span><span class="va">first100_interactions_RFmatrix_ctrl.lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterInteractions.html">FilterInteractions</a></span><span class="op">(</span></span>
<span>    matrices  <span class="op">=</span> <span class="va">interactions_RFmatrix_ctrl.lst</span>,</span>
<span>    targets    <span class="op">=</span> <span class="va">first100_targets</span>,</span>
<span>    selectionFun <span class="op">=</span> <span class="cn">NULL</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p><strong>Orientation</strong></p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oriented_first100_interactions_RFmatrix_ctrl.lst</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu"><a href="../reference/OrientateMatrix.html">OrientateMatrix</a></span><span class="op">(</span><span class="va">first100_interactions_RFmatrix_ctrl.lst</span><span class="op">)</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Preparation of second matrices list in Beaf depleted condition.
<strong>Extraction</strong>
</li>
</ol>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">interactions_RFmatrix.lst</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExtractSubmatrix.html">ExtractSubmatrix</a></span><span class="op">(</span></span>
<span>    genomicFeature         <span class="op">=</span> <span class="va">interactions.gni</span>,</span>
<span>    hicLst        <span class="op">=</span> <span class="va">HiC_HS.cmx_lst</span>,</span>
<span>    referencePoint <span class="op">=</span> <span class="st">"rf"</span>,</span>
<span>    matriceDim     <span class="op">=</span> <span class="fl">101</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p><strong>Filtration</strong></p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">first100_interactions_RFmatrix.lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterInteractions.html">FilterInteractions</a></span><span class="op">(</span></span>
<span>    matrices  <span class="op">=</span> <span class="va">interactions_RFmatrix.lst</span>,</span>
<span>    targets    <span class="op">=</span> <span class="va">first100_targets</span>,</span>
<span>    selectionFun <span class="op">=</span> <span class="cn">NULL</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p><strong>Orientation</strong></p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oriented_first100_interactions_RFmatrix.lst</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu"><a href="../reference/OrientateMatrix.html">OrientateMatrix</a></span><span class="op">(</span><span class="va">first100_interactions_RFmatrix.lst</span><span class="op">)</span></span>
<span><span class="co">#&gt; 69 matrices are oriented</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="aggregate">Aggregate<a class="anchor" aria-label="anchor" href="#aggregate"></a>
</h4>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oriented_first100_interactions_RFmatrix_ctrl.lst</span> <span class="op">=</span> </span>
<span>    <span class="fu"><a href="../reference/PrepareMtxList.html">PrepareMtxList</a></span><span class="op">(</span><span class="va">first100_interactions_RFmatrix_ctrl.lst</span>,</span>
<span>        minDist   <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>        maxDist   <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>        rm0       <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>        orientate <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; 69 matrices are oriented</span></span>
<span><span class="va">oriented_first100_interactions_RFmatrix.lst</span> <span class="op">=</span> </span>
<span>    <span class="fu"><a href="../reference/PrepareMtxList.html">PrepareMtxList</a></span><span class="op">(</span><span class="va">first100_interactions_RFmatrix.lst</span>,</span>
<span>        minDist   <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>        maxDist   <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>        rm0       <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>        orientate <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; 69 matrices are oriented</span></span>
<span></span>
<span><span class="va">diffAggreg.mtx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Aggregation.html">Aggregation</a></span><span class="op">(</span></span>
<span>    ctrlMatrices    <span class="op">=</span> <span class="va">oriented_first100_interactions_RFmatrix_ctrl.lst</span>,</span>
<span>    matrices        <span class="op">=</span> <span class="va">oriented_first100_interactions_RFmatrix.lst</span>,</span>
<span>    aggFun             <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>    diffFun            <span class="op">=</span> <span class="st">"substraction"</span>,</span>
<span>    scaleCorrection <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    correctionArea  <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span>,</span>
<span>        j <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">72</span><span class="op">:</span><span class="fl">101</span><span class="op">)</span></span>
<span>        <span class="op">)</span>,</span>
<span>    statCompare <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="tips-7">Tips<a class="anchor" aria-label="anchor" href="#tips-7"></a>
</h3>
<p>On <code>PrepareMtxList</code> function:</p>
<ol style="list-style-type: decimal">
<li>
<code>PrepareMtxList</code> acts as a one stop function to perform
value treatment and orientation correction allowing to have consistent
matrices list for both quantification and visualization process to
come.</li>
<li>If <code>rm0</code> is <code>TRUE</code> all zeros in matrices list
will be replaced by NA.</li>
<li>It is possible to filter submatrices list by minimal or maximal
distance during the aggregation function.</li>
<li>It is possible to orientate submatrices at this point or using
<code>OrientateMatrix</code> function.</li>
<li>
<code>PrepareMtxList</code> keeps former attributes matrices list
and adds new ones:
<ul>
<li>totalMatrixNumber: total number of matrices.</li>
<li>filteredMatrixNumber: number of matrices after distance
filtering.<br>
</li>
<li>minimalDistance: minimal distance between anchor and bait.<br>
</li>
<li>maximalDistance: maximal distance between anchor and bait.</li>
<li>transformationMethod: the function used to perform a per matrix data
transformation.</li>
<li>zeroRemoved: A Boolean that indicates if zeros have been replaced by
NA.</li>
</ul>
</li>
</ol>
<p>On <code>Aggregation</code> function:</p>
<ol style="list-style-type: decimal">
<li>When aggregation is performed using one sample only, use either
<code>matrices</code> or <code>ctrlMatrices</code> parameters</li>
<li>The <code>statCompare</code> may not be set <code>TRUE</code> every
time (due to memory requirement).</li>
<li>
<code>Aggregation</code> on one sample keeps former attributes of
the matrices list and add new ones:
<ul>
<li>aggregationMethod: The function applied to obtain the
aggregation.<br>
</li>
</ul>
</li>
<li>
<code>Aggregation</code> on two samples adds additional attributes:
<ul>
<li>correctedFact: The value that is added to the condition to reduce
noise. It’s computed as the median difference between condition and
control in an background area (e.g upper right corner in matrices).</li>
<li>matrices: The list of matrices.
<ul>
<li>agg: Aggregation of the condition.<br>
</li>
<li>aggCtrl: Aggregation of the control.<br>
</li>
<li>aggCorrected: Aggregation of the condition corrected with
correctedFact.</li>
<li>aggDelta: the difference between the aggregated matrix of the
condition and the aggregated matrix of the control.<br>
</li>
<li>aggCorrectedDelta: the difference between the aggregated matrix of
the condition corrected with correctedFact and the aggregated matrix of
the control.</li>
</ul>
</li>
</ul>
</li>
</ol>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="aggregations-plots">Aggregations plots<a class="anchor" aria-label="anchor" href="#aggregations-plots"></a>
</h2>
<div class="section level3">
<h3 id="preparation-of-aggregated-matrices">Preparation of aggregated matrices<a class="anchor" aria-label="anchor" href="#preparation-of-aggregated-matrices"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Control aggregation with no orientation</li>
</ol>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aggreg.mtx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Aggregation.html">Aggregation</a></span><span class="op">(</span></span>
<span>        ctrlMatrices<span class="op">=</span><span class="va">interactions_RFmatrix_ctrl.lst</span>,</span>
<span>        aggFun<span class="op">=</span><span class="st">"mean"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Control aggregation with orientation</li>
</ol>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oriented_interactions_RFmatrix_ctrl.lst</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu"><a href="../reference/OrientateMatrix.html">OrientateMatrix</a></span><span class="op">(</span><span class="va">interactions_RFmatrix_ctrl.lst</span><span class="op">)</span></span>
<span><span class="co">#&gt; 95 matrices are oriented</span></span>
<span><span class="va">orientedAggreg.mtx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Aggregation.html">Aggregation</a></span><span class="op">(</span></span>
<span>        ctrlMatrices<span class="op">=</span><span class="va">oriented_interactions_RFmatrix_ctrl.lst</span>,</span>
<span>        aggFun<span class="op">=</span><span class="st">"mean"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Differential aggregation</li>
</ol>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oriented_interactions_RFmatrix.lst</span> <span class="op">&lt;-</span> </span>
<span>    <span class="fu"><a href="../reference/OrientateMatrix.html">OrientateMatrix</a></span><span class="op">(</span><span class="va">interactions_RFmatrix.lst</span><span class="op">)</span></span>
<span><span class="co">#&gt; 95 matrices are oriented</span></span>
<span><span class="va">diffAggreg.mtx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Aggregation.html">Aggregation</a></span><span class="op">(</span></span>
<span>        ctrlMatrices    <span class="op">=</span> <span class="va">oriented_interactions_RFmatrix_ctrl.lst</span>,</span>
<span>        matrices        <span class="op">=</span> <span class="va">oriented_interactions_RFmatrix.lst</span>,</span>
<span>        aggFun          <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>        diffFun         <span class="op">=</span> <span class="st">"log2+1"</span>,</span>
<span>        scaleCorrection <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        correctionArea  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span> i<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span> , j<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">72</span><span class="op">:</span><span class="fl">101</span><span class="op">)</span> <span class="op">)</span>,</span>
<span>        statCompare     <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plots">Plots<a class="anchor" aria-label="anchor" href="#plots"></a>
</h3>
<div class="section level4">
<h4 id="simple-aggregation-plot">Simple aggregation plot:<a class="anchor" aria-label="anchor" href="#simple-aggregation-plot"></a>
</h4>
<div class="section level5">
<h5 id="with-no-orientation">With no orientation<a class="anchor" aria-label="anchor" href="#with-no-orientation"></a>
</h5>
<p><code>ggAPA</code> function creates a ggplot object
(ggplot2::geom_raster)</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggAPA.html">ggAPA</a></span><span class="op">(</span></span>
<span>        aggregatedMtx   <span class="op">=</span> <span class="va">aggreg.mtx</span>,</span>
<span>        title <span class="op">=</span> <span class="st">"APA"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-105-1.png" width="672"></p>
</div>
<div class="section level5">
<h5 id="with-orientation">With Orientation<a class="anchor" aria-label="anchor" href="#with-orientation"></a>
</h5>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggAPA.html">ggAPA</a></span><span class="op">(</span></span>
<span>        aggregatedMtx   <span class="op">=</span> <span class="va">orientedAggreg.mtx</span>,</span>
<span>        title <span class="op">=</span> <span class="st">"APA"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-106-1.png" width="672"></p>
</div>
</div>
<div class="section level4">
<h4 id="further-visualisation-parameters">Further visualisation parameters:<a class="anchor" aria-label="anchor" href="#further-visualisation-parameters"></a>
</h4>
<div class="section level5">
<h5 id="trimming-aggregated-values-for-visualisation">Trimming aggregated values for visualisation:<a class="anchor" aria-label="anchor" href="#trimming-aggregated-values-for-visualisation"></a>
</h5>
<p>It is possible to set a specific range of values of the scale, for
this remove a percentage of values using upper tail, lower tail or both
tails of the distribution.</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggAPA.html">ggAPA</a></span><span class="op">(</span></span>
<span>        aggregatedMtx      <span class="op">=</span> <span class="va">orientedAggreg.mtx</span>,</span>
<span>        title    <span class="op">=</span> <span class="st">"APA 30% trimmed on upper side"</span>,</span>
<span>        trim <span class="op">=</span> <span class="fl">30</span>,</span>
<span>        tails   <span class="op">=</span> <span class="st">"upper"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in max(unlist(bounds.num_lst[1]), na.rm = TRUE): no non-missing</span></span>
<span><span class="co">#&gt; arguments to max; returning -Inf</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-107-1.png" width="672"></p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggAPA.html">ggAPA</a></span><span class="op">(</span></span>
<span>        aggregatedMtx      <span class="op">=</span> <span class="va">orientedAggreg.mtx</span>,</span>
<span>        title    <span class="op">=</span> <span class="st">"APA 30% trimmed on upper side"</span>,</span>
<span>        trim <span class="op">=</span> <span class="fl">30</span>,</span>
<span>        tails   <span class="op">=</span> <span class="st">"lower"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in min(unlist(bounds.num_lst[2]), na.rm = TRUE): no non-missing</span></span>
<span><span class="co">#&gt; arguments to min; returning Inf</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-107-2.png" width="672"></p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggAPA.html">ggAPA</a></span><span class="op">(</span></span>
<span>        aggregatedMtx      <span class="op">=</span> <span class="va">orientedAggreg.mtx</span>,</span>
<span>        title    <span class="op">=</span> <span class="st">"APA 30% trimmed"</span>,</span>
<span>        trim <span class="op">=</span> <span class="fl">30</span>,</span>
<span>        tails   <span class="op">=</span> <span class="st">"both"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-107-3.png" width="672"></p>
</div>
<div class="section level5">
<h5 id="modifying-color-scale">Modifying color scale:<a class="anchor" aria-label="anchor" href="#modifying-color-scale"></a>
</h5>
<div class="section level6">
<h6 id="min-and-max-color-scale">Min and max color scale<a class="anchor" aria-label="anchor" href="#min-and-max-color-scale"></a>
</h6>
<p>Example of user-defined min and max color scale</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggAPA.html">ggAPA</a></span><span class="op">(</span></span>
<span>        aggregatedMtx         <span class="op">=</span> <span class="va">orientedAggreg.mtx</span>,</span>
<span>        title       <span class="op">=</span> <span class="st">"APA [0-1]"</span>,</span>
<span>        colMin <span class="op">=</span> <span class="fl">0</span>,</span>
<span>        colMax <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-108-1.png" width="672"></p>
</div>
<div class="section level6">
<h6 id="center-color-scale">Center color scale<a class="anchor" aria-label="anchor" href="#center-color-scale"></a>
</h6>
<p>Example of user-defined color scale center</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggAPA.html">ggAPA</a></span><span class="op">(</span></span>
<span>        aggregatedMtx    <span class="op">=</span> <span class="va">orientedAggreg.mtx</span>,</span>
<span>        title  <span class="op">=</span> <span class="st">"APA center on 0.2"</span>,</span>
<span>        colMid <span class="op">=</span> <span class="fl">0.5</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-109-1.png" width="672"></p>
</div>
<div class="section level6">
<h6 id="change-color-breaks">Change color breaks<a class="anchor" aria-label="anchor" href="#change-color-breaks"></a>
</h6>
<p>Examples of user-defined color breaks</p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggAPA.html">ggAPA</a></span><span class="op">(</span></span>
<span>        aggregatedMtx       <span class="op">=</span> <span class="va">orientedAggreg.mtx</span>,</span>
<span>        title     <span class="op">=</span> <span class="st">"APA [0, .25, .50, .30, .75, 1]"</span>,</span>
<span>        colBreaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0.25</span>,<span class="fl">0.5</span>,<span class="fl">0.75</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-110-1.png" width="672"></p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggAPA.html">ggAPA</a></span><span class="op">(</span></span>
<span>        aggregatedMtx       <span class="op">=</span> <span class="va">orientedAggreg.mtx</span>,</span>
<span>        title     <span class="op">=</span> <span class="st">"APA [0, .15, .20, .25, 1]"</span>,</span>
<span>        colBreaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0.15</span>,<span class="fl">0.20</span>,<span class="fl">0.25</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-110-2.png" width="672"></p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggAPA.html">ggAPA</a></span><span class="op">(</span></span>
<span>        aggregatedMtx       <span class="op">=</span> <span class="va">orientedAggreg.mtx</span>,</span>
<span>        title     <span class="op">=</span> <span class="st">"APA [0, .5, .6, .8, 1]"</span>,</span>
<span>        colBreaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0.4</span>,<span class="fl">0.5</span>,<span class="fl">0.7</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-110-3.png" width="672"></p>
</div>
<div class="section level6">
<h6 id="change-color-scale-bias">Change color scale bias<a class="anchor" aria-label="anchor" href="#change-color-scale-bias"></a>
</h6>
<p>Examples of different color scaled bias.</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggAPA.html">ggAPA</a></span><span class="op">(</span></span>
<span>        aggregatedMtx    <span class="op">=</span> <span class="va">orientedAggreg.mtx</span>,</span>
<span>        title  <span class="op">=</span> <span class="st">"APA"</span>,</span>
<span>        colorScale <span class="op">=</span> <span class="st">"density"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-111-1.png" width="672"></p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggAPA.html">ggAPA</a></span><span class="op">(</span></span>
<span>        aggregatedMtx     <span class="op">=</span> <span class="va">orientedAggreg.mtx</span>,</span>
<span>        title   <span class="op">=</span> <span class="st">"APA"</span>,</span>
<span>        bias    <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-111-2.png" width="672"></p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggAPA.html">ggAPA</a></span><span class="op">(</span></span>
<span>        aggregatedMtx     <span class="op">=</span> <span class="va">orientedAggreg.mtx</span>,</span>
<span>        title   <span class="op">=</span> <span class="st">"APA"</span>,</span>
<span>        bias    <span class="op">=</span> <span class="fl">0.5</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-111-3.png" width="672"></p>
</div>
<div class="section level6">
<h6 id="change-color">Change color<a class="anchor" aria-label="anchor" href="#change-color"></a>
</h6>
<p>Here is an option to change the color of heatmap and color of NA
values.</p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggAPA.html">ggAPA</a></span><span class="op">(</span></span>
<span>    aggregatedMtx     <span class="op">=</span> <span class="va">orientedAggreg.mtx</span>,</span>
<span>    title   <span class="op">=</span> <span class="st">"APA"</span>,</span>
<span>    colors <span class="op">=</span> <span class="fu"><a href="../reference/viridis.html">viridis</a></span><span class="op">(</span><span class="fl">6</span><span class="op">)</span>,</span>
<span>    na.value      <span class="op">=</span> <span class="st">"black"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-112-1.png" width="672"></p>
</div>
</div>
<div class="section level5">
<h5 id="blurred-visualization">Blurred visualization<a class="anchor" aria-label="anchor" href="#blurred-visualization"></a>
</h5>
<p>Option to apply a blurr on the heatmap to reduce noise.</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggAPA.html">ggAPA</a></span><span class="op">(</span></span>
<span>    aggregatedMtx           <span class="op">=</span> <span class="va">orientedAggreg.mtx</span>,</span>
<span>    title         <span class="op">=</span> <span class="st">"APA"</span>,</span>
<span>    blurPass      <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    stdev        <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>    loTri      <span class="op">=</span> <span class="cn">NA</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-113-1.png" width="672"></p>
</div>
<div class="section level5">
<h5 id="ggplot-object-modifications">ggplot object modifications<a class="anchor" aria-label="anchor" href="#ggplot-object-modifications"></a>
</h5>
<p>Since ggAPA() returns a ggplot object, it is possible to modify it
following the ggplot2 grammar</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggAPA.html">ggAPA</a></span><span class="op">(</span></span>
<span>        aggregatedMtx     <span class="op">=</span> <span class="va">orientedAggreg.mtx</span>,</span>
<span>        title   <span class="op">=</span> <span class="st">"APA"</span>,</span>
<span><span class="op">)</span> <span class="op">+</span> </span>
<span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>        title    <span class="op">=</span> <span class="st">"New title"</span>,</span>
<span>        subtitle <span class="op">=</span> <span class="st">"and subtitle"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="InDepth_files/figure-html/unnamed-chunk-114-1.png" width="672"></p>
<hr>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.0 (2024-04-24)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] HicAggR_0.99.7</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] tidyselect_1.2.1            viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt;   [3] farver_2.1.1                dplyr_1.1.4                </span></span>
<span><span class="co">#&gt;   [5] blob_1.2.4                  filelock_1.0.3             </span></span>
<span><span class="co">#&gt;   [7] fastmap_1.1.1               reshape_0.8.9              </span></span>
<span><span class="co">#&gt;   [9] BiocFileCache_2.11.2        digest_0.6.35              </span></span>
<span><span class="co">#&gt;  [11] lifecycle_1.0.4             RSQLite_2.3.6              </span></span>
<span><span class="co">#&gt;  [13] magrittr_2.0.3              compiler_4.4.0             </span></span>
<span><span class="co">#&gt;  [15] rlang_1.1.3                 sass_0.4.9                 </span></span>
<span><span class="co">#&gt;  [17] tools_4.4.0                 utf8_1.2.4                 </span></span>
<span><span class="co">#&gt;  [19] yaml_2.3.8                  data.table_1.15.4          </span></span>
<span><span class="co">#&gt;  [21] knitr_1.46                  labeling_0.4.3             </span></span>
<span><span class="co">#&gt;  [23] S4Arrays_1.3.7              bit_4.0.5                  </span></span>
<span><span class="co">#&gt;  [25] curl_5.2.1                  DelayedArray_0.29.9        </span></span>
<span><span class="co">#&gt;  [27] xml2_1.3.6                  plyr_1.8.9                 </span></span>
<span><span class="co">#&gt;  [29] abind_1.4-5                 BiocParallel_1.37.1        </span></span>
<span><span class="co">#&gt;  [31] withr_3.0.0                 purrr_1.0.2                </span></span>
<span><span class="co">#&gt;  [33] BiocGenerics_0.49.1         desc_1.4.3                 </span></span>
<span><span class="co">#&gt;  [35] grid_4.4.0                  stats4_4.4.0               </span></span>
<span><span class="co">#&gt;  [37] fansi_1.0.6                 colorspace_2.1-0           </span></span>
<span><span class="co">#&gt;  [39] Rhdf5lib_1.25.3             ggplot2_3.5.1              </span></span>
<span><span class="co">#&gt;  [41] scales_1.3.0                SummarizedExperiment_1.33.3</span></span>
<span><span class="co">#&gt;  [43] cli_3.6.2                   rmarkdown_2.26             </span></span>
<span><span class="co">#&gt;  [45] crayon_1.5.2                ragg_1.3.0                 </span></span>
<span><span class="co">#&gt;  [47] generics_0.1.3              rstudioapi_0.16.0          </span></span>
<span><span class="co">#&gt;  [49] httr_1.4.7                  DBI_1.2.2                  </span></span>
<span><span class="co">#&gt;  [51] cachem_1.0.8                rhdf5_2.47.7               </span></span>
<span><span class="co">#&gt;  [53] stringr_1.5.1               zlibbioc_1.49.3            </span></span>
<span><span class="co">#&gt;  [55] parallel_4.4.0              XVector_0.43.1             </span></span>
<span><span class="co">#&gt;  [57] matrixStats_1.3.0           vctrs_0.6.5                </span></span>
<span><span class="co">#&gt;  [59] Matrix_1.7-0                jsonlite_1.8.8             </span></span>
<span><span class="co">#&gt;  [61] IRanges_2.37.1              S4Vectors_0.41.7           </span></span>
<span><span class="co">#&gt;  [63] bit64_4.0.5                 systemfonts_1.0.6          </span></span>
<span><span class="co">#&gt;  [65] tidyr_1.3.1                 strawr_0.0.91              </span></span>
<span><span class="co">#&gt;  [67] jquerylib_0.1.4             glue_1.7.0                 </span></span>
<span><span class="co">#&gt;  [69] pkgdown_2.0.9               codetools_0.2-20           </span></span>
<span><span class="co">#&gt;  [71] stringi_1.8.3               gtable_0.3.5               </span></span>
<span><span class="co">#&gt;  [73] GenomeInfoDb_1.39.14        GenomicRanges_1.55.4       </span></span>
<span><span class="co">#&gt;  [75] UCSC.utils_0.99.7           munsell_0.5.1              </span></span>
<span><span class="co">#&gt;  [77] tibble_3.2.1                pillar_1.9.0               </span></span>
<span><span class="co">#&gt;  [79] htmltools_0.5.8.1           rhdf5filters_1.15.5        </span></span>
<span><span class="co">#&gt;  [81] GenomeInfoDbData_1.2.12     R6_2.5.1                   </span></span>
<span><span class="co">#&gt;  [83] dbplyr_2.5.0                textshaping_0.3.7          </span></span>
<span><span class="co">#&gt;  [85] kableExtra_1.4.0            evaluate_0.23              </span></span>
<span><span class="co">#&gt;  [87] lattice_0.22-6              Biobase_2.63.1             </span></span>
<span><span class="co">#&gt;  [89] highr_0.10                  backports_1.4.1            </span></span>
<span><span class="co">#&gt;  [91] memoise_2.0.1               bslib_0.7.0                </span></span>
<span><span class="co">#&gt;  [93] Rcpp_1.0.12                 InteractionSet_1.31.0      </span></span>
<span><span class="co">#&gt;  [95] svglite_2.1.3               gridExtra_2.3              </span></span>
<span><span class="co">#&gt;  [97] SparseArray_1.3.5           checkmate_2.3.1            </span></span>
<span><span class="co">#&gt;  [99] xfun_0.43                   fs_1.6.4                   </span></span>
<span><span class="co">#&gt; [101] MatrixGenerics_1.15.1       pkgconfig_2.0.3</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nicolas Chanard, David Depierre, Robel Tesfaye, Stephane Schaak, Olivier Cuvier.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
